{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figures Assessing bioinformatics software annotations: bio.tools case-study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 1 : Direct and inferred EDAM annotations for the scientific topic of Qiime2 and Vsearch, two metagenomics tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EDAM version: 1.25\n"
     ]
    }
   ],
   "source": [
    "#Python 3.13.1\n",
    "#Import library \n",
    "import pandas as pd\n",
    "import IPython\n",
    "import collections\n",
    "import pygraphviz as pgv\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "#Read csv \n",
    "dfTool = pd.read_csv(\"../Dataframes/dfTool.tsv.bz2\", sep=\"\\t\") #all Tool and toolLabel\n",
    "dfToolTopicTransitive = pd.read_csv(\"../Dataframes/dfToolTopicTransitive.tsv.bz2\", sep=\"\\t\")#Tool, topic, topicLabel inherited annotation\n",
    "dfToolOperationTransitive = pd.read_csv(\"../Dataframes/dfToolOperationTransitive.tsv.bz2\", sep=\"\\t\")# tool, operation, operationLabel inherited annotation\n",
    "\n",
    "#Configuration SPARQL end point \n",
    "endpointURL = \"http://localhost:3030/biotoolsEdam/query\"\n",
    "rdfFormat = \"turtle\"\n",
    "query=\"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT ?ontology ?versionIRI (REPLACE(STR(?versionIRI), 'http://edamontology.org/', '') AS ?versionNumber)\n",
    "WHERE {\n",
    "  ?ontology rdf:type owl:Ontology .\n",
    "  ?ontology owl:versionIRI ?versionIRI .\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+query)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "\n",
    "chebiVersion = float(results[\"results\"][\"bindings\"][0][\"versionNumber\"][\"value\"])\n",
    "print(\"EDAM version: {}\".format(chebiVersion))\n",
    "\n",
    "#Import prefix : \n",
    "prefixes = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>\n",
    "\n",
    "PREFIX bt: <https://bio.tools/>\n",
    "PREFIX biotools: <https://bio.tools/ontology/>\n",
    "PREFIX bsc: <http://bioschemas.org/>\n",
    "PREFIX bsct: <http://bioschemas.org/types/>\n",
    "PREFIX edam: <http://edamontology.org/>\n",
    "PREFIX sc: <http://schema.org/>\n",
    "PREFIX schema: <https://schema.org/>\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#Link tool\n",
    "biotoolsURI = \"https://bio.tools/\"\n",
    "biotoolsOntologyURI = \"https://bio.tools/ontology/\"\n",
    "edamURI = \"http://edamontology.org/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def addToolAndAnnotationsToGraph(toolURI, graph=None, showTopics=True, showOperations=True, showDeprecatedAnnotations=False, highlightDirectAnnotations=False):\n",
    "    \"\"\"Return a graph representing a tool and its EDAM annotations.\n",
    "\n",
    "    Keyword arguments:\n",
    "    toolURI -- the URI for the tool\n",
    "    graph -- the graph in which the tool and its annotations are added. A new graph is created if the value is None. (default: None)\n",
    "    showTopics -- should the topics annotating the tool be considered (default: True)\n",
    "    showOperations -- should the operations annotating the tool be considered (default: True)\n",
    "    showDeprecatedAnnotations -- should the deprecated topics and operations annotating the tool be considered (default: False)\n",
    "    highlightDirectAnnotations -- should the topics or operations annotated directly be highliigthed (default:False)\n",
    "    \"\"\"\n",
    "    if graph is None:\n",
    "        graph = pgv.AGraph(directed=True, rankdir=\"BT\")\n",
    "    \n",
    "    toolIdent = toolURI.replace(biotoolsURI, \"\")\n",
    "    \n",
    "    if toolURI.startswith(\"http\"):\n",
    "        toolURI = \"<\" + toolURI + \">\"\n",
    "        \n",
    "    directAnnotationColor = \"red\" if highlightDirectAnnotations else \"black\"\n",
    "    \n",
    "    conceptStyle = {}\n",
    "    conceptStyle['Tool'] = \"filled\"\n",
    "    conceptStyle['Topic'] = \"filled\"\n",
    "    conceptStyle['TopicDeprecated'] = conceptStyle['Topic']+\",dashed\"\n",
    "    conceptStyle['TopicAlternative'] = conceptStyle['Topic']+\",dotted\"\n",
    "    conceptStyle['Operation'] = \"rounded,filled\"\n",
    "    conceptStyle['OperationDeprecated'] = conceptStyle['Operation']+\",dashed\"\n",
    "    conceptStyle['OperationAlternative'] = conceptStyle['Operation']+\",dotted\"\n",
    "    \n",
    "    query = \"\"\"\n",
    "SELECT DISTINCT ?toolLabel \n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  OPTIONAL { ?tool sc:name ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?toolLabel)\n",
    "}\n",
    "\"\"\"\n",
    "    sparql = SPARQLWrapper(endpointURL)\n",
    "    sparql.setQuery(prefixes+query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.query().convert()\n",
    "    for result in results[\"results\"][\"bindings\"]:\n",
    "        #print(\"{}\\t{}\".format(toolIdent, result[\"toolLabel\"][\"value\"]))\n",
    "        if not graph.has_node(toolIdent):\n",
    "            clusterTools = graph.get_subgraph(name=\"cluster_tools\")\n",
    "            if clusterTools is None:\n",
    "                clusterTools = graph.add_subgraph(name=\"cluster_tools\", rankdir=\"same\", style=\"invis\") # style=\"invis\"\n",
    "            clusterTools.add_node(toolIdent, label=\"{}\".format(result[\"toolLabel\"][\"value\"]), shape=\"ellipse\", color=\"blue\", nodeType=\"Tool\", style=conceptStyle[\"Tool\"], fillcolor=\"#ffffff\")\n",
    "    \n",
    "    directConcepts = []\n",
    "    if showTopics:\n",
    "        directConcepts = []\n",
    "        conceptType = \"Topic\"\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?conceptURI ?conceptLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:applicationSubCategory ?conceptURI .\n",
    "  ?conceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?conceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?conceptURI rdfs:label ?cLabel }\n",
    "  BIND(COALESCE(?cLabel, \"\") AS ?conceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            directConcepts.append(result[\"conceptURI\"][\"value\"])\n",
    "            conceptIdent = result[\"conceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            #print(\"\\t{}\\t{}\".format(conceptIdent, result[\"conceptLabel\"][\"value\"]))\n",
    "            if not graph.has_node(conceptIdent):\n",
    "                graph.add_node(conceptIdent, label=\"{}\\n({})\".format(result[\"conceptLabel\"][\"value\"], conceptIdent), nodeType=conceptType, shape=\"box\", color=directAnnotationColor, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(toolIdent, conceptIdent, arrowhead=\"vee\", color=\"blue\", fontcolor=\"blue\", style=\"dashed\")\n",
    "    \n",
    "    for conceptURI in directConcepts:\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?subConceptURI ?subConceptLabel ?superConceptURI ?superConceptLabel \n",
    "WHERE {\n",
    "  VALUES ?conceptURI { <\"\"\" + conceptURI + \"\"\"> }\n",
    "  ?conceptURI rdfs:subClassOf* ?subConceptURI .\n",
    "  ?subConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?subConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?subConceptURI rdfs:label ?subLabel }\n",
    "  ?subConceptURI rdfs:subClassOf ?superConceptURI .\n",
    "  ?superConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?superConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?superConceptURI rdfs:label ?supLabel }\n",
    "  BIND(COALESCE(?subLabel, \"\") AS ?subConceptLabel)\n",
    "  BIND(COALESCE(?supLabel, \"\") AS ?superConceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        directConcepts = []\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            subConceptIdent = result[\"subConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            superConceptIdent = result[\"superConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            if not graph.has_node(subConceptIdent):\n",
    "                graph.add_node(subConceptIdent, label=\"{}\\n{}\".format(result[\"subConceptLabel\"][\"value\"], subConceptIdent), shape=\"box\", color=\"black\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            if not graph.has_node(superConceptIdent):\n",
    "                graph.add_node(superConceptIdent, label=\"{}\\n{}\".format(result[\"superConceptLabel\"][\"value\"], superConceptIdent), shape=\"box\", color=\"black\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(subConceptIdent, superConceptIdent, arrowhead=\"onormal\")\n",
    "            \n",
    "            \n",
    "    if showOperations:\n",
    "        directConcepts = []\n",
    "        conceptType = \"Operation\"\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?conceptURI ?conceptLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:featureList ?conceptURI .\n",
    "  ?conceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?conceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?conceptURI rdfs:label ?cLabel }\n",
    "  BIND(COALESCE(?cLabel, \"\") AS ?conceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            directConcepts.append(result[\"conceptURI\"][\"value\"])\n",
    "            conceptIdent = result[\"conceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            #print(\"\\t{}\\t{}\".format(conceptIdent, result[\"conceptLabel\"][\"value\"]))\n",
    "            if not graph.has_node(conceptIdent):\n",
    "                graph.add_node(conceptIdent, label=\"{}\\n({})\".format(result[\"conceptLabel\"][\"value\"], conceptIdent), nodeType=conceptType, shape=\"box\", color=directAnnotationColor, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(toolIdent, conceptIdent, arrowhead=\"vee\", color=\"blue\", fontcolor=\"blue\", style=\"dashed\")\n",
    "    \n",
    "    for conceptURI in directConcepts:\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?subConceptURI ?subConceptLabel ?superConceptURI ?superConceptLabel \n",
    "WHERE {\n",
    "  VALUES ?conceptURI { <\"\"\" + conceptURI + \"\"\"> }\n",
    "  ?conceptURI rdfs:subClassOf* ?subConceptURI .\n",
    "  ?subConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?subConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?subConceptURI rdfs:label ?subLabel }\n",
    "  ?subConceptURI rdfs:subClassOf ?superConceptURI .\n",
    "  ?superConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?superConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?superConceptURI rdfs:label ?supLabel }\n",
    "  BIND(COALESCE(?subLabel, \"\") AS ?subConceptLabel)\n",
    "  BIND(COALESCE(?supLabel, \"\") AS ?superConceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        directConcepts = []\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            subConceptIdent = result[\"subConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            superConceptIdent = result[\"superConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            if not graph.has_node(subConceptIdent):\n",
    "                graph.add_node(subConceptIdent, label=\"{}\\n{}\".format(result[\"subConceptLabel\"][\"value\"], subConceptIdent), shape=\"box\", color=\"black\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            if not graph.has_node(superConceptIdent):\n",
    "                graph.add_node(superConceptIdent, label=\"{}\\n{}\".format(result[\"superConceptLabel\"][\"value\"], superConceptIdent), shape=\"box\", color=\"black\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(subConceptIdent, superConceptIdent, arrowhead=\"onormal\")\n",
    "    \n",
    "    if showDeprecatedAnnotations:\n",
    "        #conceptType = \"TopicDeprecated\"\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?conceptURI ?conceptLabel ?conceptAlternative ?conceptAlternativeLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:applicationSubCategory ?conceptURI .\n",
    "  #?conceptURI rdf:type owl:Class .\n",
    "  { ?conceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated true }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated \"true\" }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated \"True\" }\n",
    "  OPTIONAL { ?conceptURI rdfs:label ?cLabel }\n",
    "  BIND(COALESCE(?cLabel, \"\") AS ?conceptLabel)\n",
    "  \n",
    "  OPTIONAL {\n",
    "    ?conceptURI oboInOwl:consider ?conceptAlternative .\n",
    "    OPTIONAL { ?conceptAlternative rdfs:label ?caLabel }\n",
    "    BIND(COALESCE(?caLabel, \"\") AS ?conceptAlternativeLabel)\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            conceptType = \"TopicDeprecated\"\n",
    "            directConcepts.append(result[\"conceptURI\"][\"value\"])\n",
    "            conceptIdent = result[\"conceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            #print(\"\\t{}\\t{}\".format(conceptIdent, result[\"conceptLabel\"][\"value\"]))\n",
    "            if not graph.has_node(conceptIdent):\n",
    "                graph.add_node(conceptIdent, label=\"{}\\n({})\".format(result[\"conceptLabel\"][\"value\"], conceptIdent), nodeType=conceptType, shape=\"box\", color=\"grey\", style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(toolIdent, conceptIdent, arrowhead=\"vee\", color=\"grey\", fontcolor=\"grey\", style=\"dashed\")\n",
    "            if \"conceptAlternative\" in result.keys():\n",
    "                conceptType = \"TopicAlternative\"\n",
    "                alternativeIdent = result[\"conceptAlternative\"][\"value\"].replace(edamURI, \"\")\n",
    "                if not graph.has_node(alternativeIdent):\n",
    "                    graph.add_node(alternativeIdent, label=\"{}\\n({})\".format(result[\"conceptAlternativeLabel\"][\"value\"], alternativeIdent), nodeType=conceptType, shape=\"box\", color=\"grey\", style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                    queryHierarchy = \"\"\"\n",
    "SELECT DISTINCT ?subConceptURI ?subConceptLabel ?superConceptURI ?superConceptLabel \n",
    "WHERE {\n",
    "  VALUES ?conceptURI { <\"\"\" + alternativeURI + \"\"\"> }\n",
    "  ?conceptURI rdfs:subClassOf* ?subConceptURI .\n",
    "  ?subConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?subConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?subConceptURI rdfs:label ?subLabel }\n",
    "  ?subConceptURI rdfs:subClassOf ?superConceptURI .\n",
    "  ?superConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?superConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?superConceptURI rdfs:label ?supLabel }\n",
    "  BIND(COALESCE(?subLabel, \"\") AS ?subConceptLabel)\n",
    "  BIND(COALESCE(?supLabel, \"\") AS ?superConceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "                    sparqlHierarchy = SPARQLWrapper(endpointURL)\n",
    "                    sparqlHierarchy.setQuery(prefixes+queryHierarchy)\n",
    "                    sparqlHierarchy.setReturnFormat(JSON)\n",
    "                    resultsHierarchy = sparqlHierarchy.query().convert()\n",
    "                    for resultHierarchy in resultsHierarchy[\"results\"][\"bindings\"]:\n",
    "                        subConceptIdent = resultHierarchy[\"subConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "                        superConceptIdent = resultHierarchy[\"superConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "                        if not graph.has_node(subConceptIdent):\n",
    "                            graph.add_node(subConceptIdent, label=\"{}\\n{}\".format(resultHierarchy[\"subConceptLabel\"][\"value\"], subConceptIdent), shape=\"box\", color=\"grey\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                        if not graph.has_node(superConceptIdent):\n",
    "                            graph.add_node(superConceptIdent, label=\"{}\\n{}\".format(resultHierarchy[\"superConceptLabel\"][\"value\"], superConceptIdent), shape=\"box\", color=\"grey\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                        if not graph.has_edge(subConceptIdent, superConceptIdent):\n",
    "                            graph.add_edge(subConceptIdent, superConceptIdent, arrowhead=\"onormal\", color=\"grey\", style=\"dotted\")\n",
    "                \n",
    "                graph.add_edge(conceptIdent, alternativeIdent, arrowhead=\"vee\", color=\"grey\", fontcolor=\"grey\", style=\"dotted\")\n",
    "    \n",
    "        \n",
    "        #conceptType = \"OperationDeprecated\"\n",
    "        query = \"\"\"\n",
    "SELECT DISTINCT ?conceptURI ?conceptLabel ?conceptAlternative ?conceptAlternativeLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:featureList ?conceptURI .\n",
    "  #?conceptURI rdf:type owl:Class .\n",
    "  { ?conceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated true }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated \"true\" }\n",
    "  UNION\n",
    "  { ?conceptURI owl:deprecated \"True\" }\n",
    "  OPTIONAL { ?conceptURI rdfs:label ?cLabel }\n",
    "  BIND(COALESCE(?cLabel, \"\") AS ?conceptLabel)\n",
    "  \n",
    "  OPTIONAL {\n",
    "    ?conceptURI oboInOwl:consider ?conceptAlternative .\n",
    "    OPTIONAL { ?conceptAlternative rdfs:label ?caLabel }\n",
    "    BIND(COALESCE(?caLabel, \"\") AS ?conceptAlternativeLabel)\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "        sparql = SPARQLWrapper(endpointURL)\n",
    "        sparql.setQuery(prefixes+query)\n",
    "        sparql.setReturnFormat(JSON)\n",
    "        results = sparql.query().convert()\n",
    "        for result in results[\"results\"][\"bindings\"]:\n",
    "            conceptType = \"OperationDeprecated\"\n",
    "            directConcepts.append(result[\"conceptURI\"][\"value\"])\n",
    "            conceptIdent = result[\"conceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "            #print(\"\\t{}\\t{}\".format(conceptIdent, result[\"conceptLabel\"][\"value\"]))\n",
    "            if not graph.has_node(conceptIdent):\n",
    "                graph.add_node(conceptIdent, label=\"{}\\n({})\".format(result[\"conceptLabel\"][\"value\"], conceptIdent), nodeType=conceptType, shape=\"box\", color=\"grey\", style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "            graph.add_edge(toolIdent, conceptIdent, arrowhead=\"vee\", color=\"grey\", fontcolor=\"grey\", style=\"dashed\")\n",
    "            if \"conceptAlternative\" in result.keys():\n",
    "                conceptType = \"OperationAlternative\"\n",
    "                alternativeURI = result[\"conceptAlternative\"][\"value\"]\n",
    "                alternativeIdent = alternativeURI.replace(edamURI, \"\")\n",
    "                if not graph.has_node(alternativeIdent):\n",
    "                    graph.add_node(alternativeIdent, label=\"{}\\n({})\".format(result[\"conceptAlternativeLabel\"][\"value\"], alternativeIdent), nodeType=conceptType, shape=\"box\", color=\"grey\", style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                    queryHierarchy = \"\"\"\n",
    "SELECT DISTINCT ?subConceptURI ?subConceptLabel ?superConceptURI ?superConceptLabel \n",
    "WHERE {\n",
    "  VALUES ?conceptURI { <\"\"\" + alternativeURI + \"\"\"> }\n",
    "  ?conceptURI rdfs:subClassOf* ?subConceptURI .\n",
    "  ?subConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?subConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?subConceptURI rdfs:label ?subLabel }\n",
    "  ?subConceptURI rdfs:subClassOf ?superConceptURI .\n",
    "  ?superConceptURI rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?superConceptURI rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?superConceptURI rdfs:label ?supLabel }\n",
    "  BIND(COALESCE(?subLabel, \"\") AS ?subConceptLabel)\n",
    "  BIND(COALESCE(?supLabel, \"\") AS ?superConceptLabel)\n",
    "}\n",
    "\"\"\"\n",
    "                    sparqlHierarchy = SPARQLWrapper(endpointURL)\n",
    "                    sparqlHierarchy.setQuery(prefixes+queryHierarchy)\n",
    "                    sparqlHierarchy.setReturnFormat(JSON)\n",
    "                    resultsHierarchy = sparqlHierarchy.query().convert()\n",
    "                    for resultHierarchy in resultsHierarchy[\"results\"][\"bindings\"]:\n",
    "                        subConceptIdent = resultHierarchy[\"subConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "                        superConceptIdent = resultHierarchy[\"superConceptURI\"][\"value\"].replace(edamURI, \"\")\n",
    "                        if not graph.has_node(subConceptIdent):\n",
    "                            graph.add_node(subConceptIdent, label=\"{}\\n{}\".format(resultHierarchy[\"subConceptLabel\"][\"value\"], subConceptIdent), shape=\"box\", color=\"grey\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                        if not graph.has_node(superConceptIdent):\n",
    "                            graph.add_node(superConceptIdent, label=\"{}\\n{}\".format(resultHierarchy[\"superConceptLabel\"][\"value\"], superConceptIdent), shape=\"box\", color=\"grey\", nodeType=conceptType, style=conceptStyle[conceptType], fillcolor=\"#ffffff\")\n",
    "                        if not graph.has_edge(subConceptIdent, superConceptIdent):\n",
    "                            graph.add_edge(subConceptIdent, superConceptIdent, arrowhead=\"onormal\", color=\"grey\", style=\"dotted\")\n",
    "                graph.add_edge(conceptIdent, alternativeIdent, arrowhead=\"vee\", color=\"grey\", fontcolor=\"grey\", style=\"dotted\")\n",
    "    \n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tool_url(tool_name: str) -> str:\n",
    "    \"\"\"Returns the full bio.tools URL for a given tool name, ensuring the correct format.\"\"\"\n",
    "    if tool_name.startswith(\"https://bio.tools/\"):\n",
    "        return tool_name\n",
    "    return f\"https://bio.tools/{tool_name}\"\n",
    "\n",
    "\n",
    "def getToolTopics(tool):\n",
    "    \"\"\"\n",
    "    Returns a list of (topic, topicLabel) tuples for the given tool.\n",
    "    \"\"\"\n",
    "    filtered = dfToolTopicTransitive[dfToolTopicTransitive['tool'] == tool]\n",
    "    return list(filtered[['topic', 'topicLabel']].itertuples(index=False, name=None))\n",
    "\n",
    "\n",
    "def getToolOperations(tool):\n",
    "    \"\"\"\n",
    "    Returns a list of (operation, operationLabel) tuples for the given tool.\n",
    "    \"\"\"\n",
    "    filtered = dfToolOperationTransitive[dfToolOperationTransitive['tool'] == tool]\n",
    "    return list(filtered[['operation', 'operationLabel']].itertuples(index=False, name=None))\n",
    "\n",
    "\n",
    "def getToolsCommonTopics(listToolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the topics associated to all the tools of a list.\n",
    "\n",
    "    Keyword arguments:\n",
    "    listToolURI -- list of the URIs for the tools\n",
    "    transitive -- also consider the ancestors of the topics directly associated to a tool (default: False)\n",
    "    \"\"\"\n",
    "    commonConcepts = []\n",
    "    if len(listToolURI) > 0:\n",
    "        commonConcepts = set(getToolTopics(listToolURI[0]))\n",
    "    for toolURI in listToolURI[1:]:\n",
    "        currentConcepts = set(getToolTopics(toolURI))\n",
    "        commonConcepts = commonConcepts.intersection(currentConcepts)\n",
    "    return list(commonConcepts)\n",
    "\n",
    "\n",
    "def getToolsCommonOperations(listToolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the operations associated to all the tools of a list.\n",
    "\n",
    "    Keyword arguments:\n",
    "    listToolURI -- list of the URIs for the tools\n",
    "    transitive -- also consider the ancestors of the operations directly associated to a tool (default: False)\n",
    "    \"\"\"\n",
    "    commonConcepts = []\n",
    "    if len(listToolURI) > 0:\n",
    "        commonConcepts = set(getToolOperations(listToolURI[0]))\n",
    "    for toolURI in listToolURI[1:]:\n",
    "        currentConcepts = set(getToolOperations(toolURI))\n",
    "        commonConcepts = commonConcepts.intersection(currentConcepts)\n",
    "    return list(commonConcepts)\n",
    "\n",
    "\n",
    "def colorGraphNodesAccordingToScore(graph, dictTopicScore, dictOperationScore, color=\"red\"):\n",
    "    \"\"\"Modify nodes color according to a score associated to the node. Nodes that are not associated to a score are unaffected.\n",
    "\n",
    "    Keyword arguments:\n",
    "    graph -- the graph to be modified\n",
    "    dictTopicScore -- a dictionary {nodeIdent -> score} for topic nodes\n",
    "    dictOperationScore -- a dictionary {nodeIdent -> score} for operation nodes\n",
    "    color -- name of the color that will vary according to scores. Possible values are \"red\", \"green\" or \"blue\" (default: \"red\")\n",
    "    \"\"\"\n",
    "    topicMaxValue = max(dictTopicScore.values(), default=0)\n",
    "    operationMaxValue = max(dictOperationScore.values(), default=0)\n",
    "    for currentNode in graph.nodes_iter():\n",
    "        if \"nodeType\" in currentNode.attr.keys():\n",
    "            currentNodeType = currentNode.attr['nodeType']\n",
    "            #currentNode.attr['style'] = conceptStyle[currentNodeType]\n",
    "            if currentNodeType == \"Topic\":\n",
    "                if currentNode in dictTopicScore.keys():\n",
    "                    graph.get_node(currentNode).attr['fillcolor'] = getScoreColorRGB(dictTopicScore[currentNode], topicMaxValue, color=color)\n",
    "            elif currentNodeType == \"Operation\":\n",
    "                if currentNode in dictOperationScore.keys():\n",
    "                    graph.get_node(currentNode).attr['fillcolor'] = getScoreColorRGB(dictOperationScore[currentNode], operationMaxValue, color=color)\n",
    "\n",
    "\n",
    "def addToolsAndAnnotationsToGraph(listToolURI, graph=None, showTopics=True, showOperations=True, highlightDirectAnnotations=False, highlightIntersection=False):\n",
    "    \"\"\"Return a graph representing tools and their EDAM annotations.\n",
    "    \n",
    "    Keyword arguments:\n",
    "      listToolURI -- list of the URIs for the tools\n",
    "      graph -- the graph in which the tool and its annotations are added. A new graph is created if the value is None.\n",
    "      showTopics -- should the topics annotating the tool be considered\n",
    "      showOperations -- should the operations annotating the tool be considered\n",
    "      highlightDirectAnnotations -- should the topics or operations annotated directly be highlighted\n",
    "      highlightIntersection -- should the common topics or operations common to all the tools be highlighted\n",
    "    \"\"\"\n",
    "    if graph is None:\n",
    "        graph = pgv.AGraph(directed=True, rankdir=\"BT\")\n",
    "    \n",
    "    for toolURI in listToolURI:\n",
    "        addToolAndAnnotationsToGraph(\n",
    "            toolURI, \n",
    "            graph=graph, \n",
    "            showTopics=showTopics, \n",
    "            showOperations=showOperations, \n",
    "            highlightDirectAnnotations=highlightDirectAnnotations\n",
    "        )\n",
    "    \n",
    "    if highlightIntersection:\n",
    "        commonConcepts = []\n",
    "        if showTopics:\n",
    "            commonConcepts += getToolsCommonTopics(listToolURI, transitive=True)\n",
    "        if showOperations:\n",
    "            commonConcepts += getToolsCommonOperations(listToolURI, transitive=True)\n",
    "    \n",
    "        for (currentConcept, currentLabel) in commonConcepts:\n",
    "            currentConceptIdent = currentConcept.replace(edamURI, \"\")\n",
    "            node = graph.get_node(currentConceptIdent)\n",
    "            node.attr['color'] = \"red\"\n",
    "            node.attr['penwidth'] = \"3\"              # Thicker border\n",
    "            node.attr['style'] = \"filled,bold\"         # Bold border and filled style\n",
    "            node.attr['fillcolor'] = \"red\"             # Red fill to highlight\n",
    "            node.attr['highlightIntersection'] = \"True\"  # Mark this node as highlighted\n",
    "    \n",
    "    return graph\n",
    "\n",
    "\n",
    "def getScoreColorRGB(scoreValue, scoreMaxValue, color=\"red\"):\n",
    "    \"\"\"Return the RGB color (in hex) associated to a score, varying from white (0 score)\n",
    "    to a target color at the maximum score.\n",
    "\n",
    "    Keyword arguments:\n",
    "    scoreValue -- the score to consider (positive number)\n",
    "    scoreMaxValue -- the maximum value for the score (positive number, >= scoreValue)\n",
    "    color -- name of the color for the gradient. Possible values are:\n",
    "             \"red\", \"green\", \"blue\", \"orange\", \"yellow\", \"pink\", \"grey\".\n",
    "             (default: \"red\")\n",
    "    \"\"\"\n",
    "    # Avoid division by 0.\n",
    "    scoreMaxValue = max(1, scoreMaxValue)\n",
    "    fraction = scoreValue / scoreMaxValue\n",
    "\n",
    "    # Define the target color values.\n",
    "    if color == \"red\":\n",
    "        target = (255, 0, 0)\n",
    "    elif color == \"green\":\n",
    "        target = (0, 255, 0)\n",
    "    elif color == \"blue\":\n",
    "        target = (0, 0, 255)\n",
    "    elif color == \"orange\":\n",
    "        target = (255, 165, 0)\n",
    "    elif color == \"yellow\":\n",
    "        target = (255, 255, 0)\n",
    "    elif color == \"pink\":\n",
    "        target = (255, 192, 203)  # Light pink (alternative: hot pink (255,105,180))\n",
    "    elif color == \"grey\":\n",
    "        target = (128, 128, 128)\n",
    "    else:\n",
    "        return \"#ffffff\"\n",
    "\n",
    "    # Interpolate from white (255,255,255) to the target color.\n",
    "    r = int(255 - fraction * (255 - target[0]))\n",
    "    g = int(255 - fraction * (255 - target[1]))\n",
    "    b = int(255 - fraction * (255 - target[2]))\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(r, g, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"626pt\" height=\"599pt\" viewBox=\"0.00 0.00 626.00 599.00\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 595)\">\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-595 622,-595 622,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_tools</title>\n",
       "</g>\n",
       "<!-- qiime2 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>qiime2</title>\n",
       "<ellipse fill=\"#ffffff\" stroke=\"blue\" cx=\"305.5\" cy=\"-34\" rx=\"45.21\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"305.5\" y=\"-28.95\" font-family=\"Times,serif\" font-size=\"14.00\">QIIME 2</text>\n",
       "</g>\n",
       "<!-- topic_3941 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>topic_3941</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"376.75,-129 248.25,-129 248.25,-88 376.75,-88 376.75,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"312.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Metatranscriptomics</text>\n",
       "<text text-anchor=\"middle\" x=\"312.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3941)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3941 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3941</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M307.16,-52.18C307.86,-59.49 308.71,-68.25 309.52,-76.62\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"310.46,-86.41 305.02,-76.88 310.1,-82.64 309.5,-76.45 309.5,-76.45 309.5,-76.45 310.1,-82.64 313.98,-76.02 310.46,-86.41\"/>\n",
       "</g>\n",
       "<!-- topic_3697 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>topic_3697</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"548.75,-129 432.25,-129 432.25,-88 548.75,-88 548.75,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"490.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Microbial ecology</text>\n",
       "<text text-anchor=\"middle\" x=\"490.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3697)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3697 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3697</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M336.88,-47.3C362.17,-57.21 398.48,-71.44 429.45,-83.58\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"438.59,-87.16 427.64,-87.7 435.07,-85.78 429.28,-83.51 429.28,-83.51 429.28,-83.51 435.07,-85.78 430.93,-79.32 438.59,-87.16\"/>\n",
       "</g>\n",
       "<!-- topic_3174 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>topic_3174</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"230,-129 133,-129 133,-88 230,-88 230,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Metagenomics</text>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3174)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3174 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3174</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M280.7,-49.5C265.25,-58.53 244.9,-70.43 226.65,-81.1\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"218.11,-86.1 224.47,-77.16 221.37,-84.19 226.74,-81.05 226.74,-81.05 226.74,-81.05 221.37,-84.19 229.01,-84.93 218.11,-86.1\"/>\n",
       "</g>\n",
       "<!-- topic_0084 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>topic_0084</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"446.38,-360 360.62,-360 360.62,-319 446.38,-319 446.38,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Phylogeny</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_0084)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_0084 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_0084</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M336.62,-47.29C354.12,-55.92 374.58,-69.3 385.5,-88 426.22,-157.71 417.45,-257.62 409.48,-307.58\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"407.81,-317.39 405.06,-306.78 408.45,-313.66 409.49,-307.53 409.49,-307.53 409.49,-307.53 408.45,-313.66 413.93,-308.29 407.81,-317.39\"/>\n",
       "</g>\n",
       "<!-- vsearch -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>vsearch</title>\n",
       "<ellipse fill=\"#ffffff\" stroke=\"blue\" cx=\"188.5\" cy=\"-34\" rx=\"53.91\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"188.5\" y=\"-28.95\" font-family=\"Times,serif\" font-size=\"14.00\">VSEARCH</text>\n",
       "</g>\n",
       "<!-- vsearch&#45;&gt;topic_3174 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>vsearch-&gt;topic_3174</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M186.84,-52.18C186.17,-59.18 185.36,-67.51 184.58,-75.55\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"183.64,-85.35 180.12,-74.96 184,-81.59 184.6,-75.4 184.6,-75.4 184.6,-75.4 184,-81.59 189.08,-75.83 183.64,-85.35\"/>\n",
       "</g>\n",
       "<!-- topic_0080 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>topic_0080</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"115,-129 0,-129 0,-88 115,-88 115,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"57.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Sequence analysis</text>\n",
       "<text text-anchor=\"middle\" x=\"57.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_0080)</text>\n",
       "</g>\n",
       "<!-- vsearch&#45;&gt;topic_0080 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>vsearch-&gt;topic_0080</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M161.36,-50.02C144.56,-59.32 122.57,-71.49 103.14,-82.24\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"94.63,-86.95 101.2,-78.17 97.93,-85.12 103.37,-82.11 103.37,-82.11 103.37,-82.11 97.93,-85.12 105.55,-86.05 94.63,-86.95\"/>\n",
       "</g>\n",
       "<!-- topic_0203 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>topic_0203</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"346.5,-283 240.5,-283 240.5,-242 346.5,-242 346.5,-283\"/>\n",
       "<text text-anchor=\"middle\" x=\"293.5\" y=\"-265.7\" font-family=\"Times,serif\" font-size=\"14.00\">Gene expression</text>\n",
       "<text text-anchor=\"middle\" x=\"293.5\" y=\"-249.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0203</text>\n",
       "</g>\n",
       "<!-- topic_3941&#45;&gt;topic_0203 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>topic_3941-&gt;topic_0203</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M318.47,-129.36C323.48,-149.07 328.89,-179.99 322.5,-206 320.36,-214.74 316.57,-223.62 312.42,-231.65\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"309.41,-229.86 307.61,-240.3 315.53,-233.26 309.41,-229.86\"/>\n",
       "</g>\n",
       "<!-- topic_3308 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>topic_3308</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"313.75,-206 209.25,-206 209.25,-165 313.75,-165 313.75,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"261.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Transcriptomics</text>\n",
       "<text text-anchor=\"middle\" x=\"261.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3308</text>\n",
       "</g>\n",
       "<!-- topic_3941&#45;&gt;topic_3308 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>topic_3941-&gt;topic_3308</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M299.1,-129.21C293.7,-137.15 287.35,-146.49 281.42,-155.2\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"278.57,-153.18 275.84,-163.41 284.36,-157.11 278.57,-153.18\"/>\n",
       "</g>\n",
       "<!-- topic_0610 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>topic_0610</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"508.88,-206 432.12,-206 432.12,-165 508.88,-165 508.88,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"470.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Ecology</text>\n",
       "<text text-anchor=\"middle\" x=\"470.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0610</text>\n",
       "</g>\n",
       "<!-- topic_3697&#45;&gt;topic_0610 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>topic_3697-&gt;topic_0610</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M485.25,-129.21C483.33,-136.4 481.1,-144.74 478.97,-152.72\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"475.65,-151.59 476.46,-162.15 482.42,-153.39 475.65,-151.59\"/>\n",
       "</g>\n",
       "<!-- topic_3301 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>topic_3301</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"618,-206 527,-206 527,-165 618,-165 618,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"572.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Microbiology</text>\n",
       "<text text-anchor=\"middle\" x=\"572.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3301</text>\n",
       "</g>\n",
       "<!-- topic_3697&#45;&gt;topic_3301 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>topic_3697-&gt;topic_3301</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M512.04,-129.21C521.39,-137.75 532.48,-147.9 542.6,-157.16\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"539.96,-159.48 549.7,-163.64 544.68,-154.31 539.96,-159.48\"/>\n",
       "</g>\n",
       "<!-- topic_3174&#45;&gt;topic_0610 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>topic_3174-&gt;topic_0610</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M231.4,-126.59C234.13,-127.43 236.85,-128.24 239.5,-129 301.02,-146.69 372.85,-163.33 419.88,-173.69\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"418.99,-177.08 429.51,-175.8 420.49,-170.24 418.99,-177.08\"/>\n",
       "</g>\n",
       "<!-- topic_0622 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>topic_0622</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"222.88,-283 146.12,-283 146.12,-242 222.88,-242 222.88,-283\"/>\n",
       "<text text-anchor=\"middle\" x=\"184.5\" y=\"-265.7\" font-family=\"Times,serif\" font-size=\"14.00\">Genomics</text>\n",
       "<text text-anchor=\"middle\" x=\"184.5\" y=\"-249.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0622</text>\n",
       "</g>\n",
       "<!-- topic_3174&#45;&gt;topic_0622 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>topic_3174-&gt;topic_0622</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M181.91,-130.32C182.42,-155.9 183.28,-199.58 183.87,-229.46\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"180.37,-229.49 184.06,-239.42 187.37,-229.35 180.37,-229.49\"/>\n",
       "</g>\n",
       "<!-- topic_3307 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>topic_3307</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"225.5,-437 83.5,-437 83.5,-396 225.5,-396 225.5,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Computational biology</text>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3307</text>\n",
       "</g>\n",
       "<!-- topic_0084&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>topic_0084-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M360.21,-356.73C356.93,-357.87 353.67,-358.97 350.5,-360 313.62,-371.99 272.37,-383.89 237.26,-393.58\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"236.75,-390.09 228.04,-396.11 238.61,-396.84 236.75,-390.09\"/>\n",
       "</g>\n",
       "<!-- topic_3299 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>topic_3299</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"469.25,-437 337.75,-437 337.75,-396 469.25,-396 469.25,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Evolutionary biology</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3299</text>\n",
       "</g>\n",
       "<!-- topic_0084&#45;&gt;topic_3299 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>topic_0084-&gt;topic_3299</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M403.5,-360.21C403.5,-367.55 403.5,-376.09 403.5,-384.22\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"400,-384.15 403.5,-394.15 407,-384.15 400,-384.15\"/>\n",
       "</g>\n",
       "<!-- topic_3321 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>topic_3321</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"341.62,-360 221.38,-360 221.38,-319 341.62,-319 341.62,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Molecular genetics</text>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3321</text>\n",
       "</g>\n",
       "<!-- topic_0203&#45;&gt;topic_3321 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>topic_0203-&gt;topic_3321</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M290.35,-283.21C289.16,-290.64 287.77,-299.29 286.46,-307.51\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"283.04,-306.74 284.91,-317.17 289.95,-307.85 283.04,-306.74\"/>\n",
       "</g>\n",
       "<!-- topic_3308&#45;&gt;topic_0203 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>topic_3308-&gt;topic_0203</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M269.91,-206.21C273.19,-213.9 277.02,-222.89 280.64,-231.35\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"277.3,-232.45 284.44,-240.27 283.74,-229.7 277.3,-232.45\"/>\n",
       "</g>\n",
       "<!-- topic_3308&#45;&gt;topic_0622 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>topic_3308-&gt;topic_0622</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M241.27,-206.21C232.95,-214.31 223.14,-223.87 214.04,-232.73\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"211.75,-230.07 207.03,-239.56 216.63,-235.09 211.75,-230.07\"/>\n",
       "</g>\n",
       "<!-- topic_3070 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>topic_3070</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"441.88,-514 365.12,-514 365.12,-473 441.88,-473 441.88,-514\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-496.7\" font-family=\"Times,serif\" font-size=\"14.00\">Biology</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-480.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3070</text>\n",
       "</g>\n",
       "<!-- topic_0610&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>topic_0610-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M477.52,-207.4C491.28,-252.29 517.28,-360.07 478.5,-437 472.58,-448.75 462.87,-458.63 452.37,-466.64\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"450.42,-463.73 444.25,-472.33 454.44,-469.46 450.42,-463.73\"/>\n",
       "</g>\n",
       "<!-- topic_3301&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>topic_3301-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M571.62,-206.3C568.65,-251.79 555.69,-365.36 499.5,-437 487.5,-452.3 470.02,-464.17 453.25,-472.93\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"451.75,-469.77 444.32,-477.32 454.84,-476.05 451.75,-469.77\"/>\n",
       "</g>\n",
       "<!-- topic_3391 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>topic_3391</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"90.88,-360 14.12,-360 14.12,-319 90.88,-319 90.88,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"52.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Omics</text>\n",
       "<text text-anchor=\"middle\" x=\"52.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3391</text>\n",
       "</g>\n",
       "<!-- topic_0622&#45;&gt;topic_3391 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>topic_0622-&gt;topic_3391</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M147.73,-284.39C132.57,-293.01 114.83,-303.09 98.8,-312.2\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"97.35,-308.99 90.39,-316.97 100.81,-315.08 97.35,-308.99\"/>\n",
       "</g>\n",
       "<!-- topic_0003 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>topic_0003</title>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"3\" points=\"192.88,-591 116.12,-591 116.12,-550 192.88,-550 192.88,-591\"/>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-573.7\" font-family=\"Times,serif\" font-size=\"14.00\">Topic</text>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-557.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0003</text>\n",
       "</g>\n",
       "<!-- topic_3307&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>topic_3307-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M154.5,-438.32C154.5,-463.9 154.5,-507.58 154.5,-537.46\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"151,-537.42 154.5,-547.42 158,-537.42 151,-537.42\"/>\n",
       "</g>\n",
       "<!-- topic_3299&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>topic_3299-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M403.5,-437.21C403.5,-444.24 403.5,-452.36 403.5,-460.17\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"400,-460.1 403.5,-470.1 407,-460.1 400,-460.1\"/>\n",
       "</g>\n",
       "<!-- topic_3321&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>topic_3321-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M247.8,-360.4C232.95,-369.17 215.29,-379.6 199.35,-389.01\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"197.82,-385.85 190.99,-393.95 201.38,-391.88 197.82,-385.85\"/>\n",
       "</g>\n",
       "<!-- topic_3053 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>topic_3053</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"black\" points=\"319.88,-437 243.12,-437 243.12,-396 319.88,-396 319.88,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Genetics</text>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3053</text>\n",
       "</g>\n",
       "<!-- topic_3321&#45;&gt;topic_3053 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>topic_3321-&gt;topic_3053</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M281.5,-360.21C281.5,-367.55 281.5,-376.09 281.5,-384.22\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"278,-384.15 281.5,-394.15 285,-384.15 278,-384.15\"/>\n",
       "</g>\n",
       "<!-- topic_3053&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>topic_3053-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M313.87,-437.4C328,-446.09 344.78,-456.4 359.98,-465.75\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"358.05,-468.67 368.41,-470.93 361.72,-462.71 358.05,-468.67\"/>\n",
       "</g>\n",
       "<!-- topic_3070&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>topic_3070-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M363.77,-506.47C320.68,-519.45 251.79,-540.2 204.91,-554.32\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"204.1,-550.9 195.54,-557.14 206.12,-557.61 204.1,-550.9\"/>\n",
       "</g>\n",
       "<!-- topic_3391&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>topic_3391-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M55.29,-361.41C58.41,-381.36 64.4,-411.98 74.5,-437 89.57,-474.34 114.41,-513.37 132.42,-539.31\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"129.46,-541.19 138.08,-547.35 135.18,-537.15 129.46,-541.19\"/>\n",
       "</g>\n",
       "<!-- topic_0080&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>topic_0080-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M63.72,-129.11C80.09,-180.77 124.53,-320.96 144.52,-384.02\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"141.14,-384.94 147.5,-393.41 147.81,-382.82 141.14,-384.94\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tools = [\"qiime2\",\"vsearch\"]\n",
    "nbTools = len(dfTool)\n",
    "\n",
    "fileFormat = \"png\"\n",
    "\n",
    "listToolURI = [biotoolsURI+x for x in tools]\n",
    "commonGraph = addToolsAndAnnotationsToGraph(listToolURI, graph=None, showTopics=True, showOperations=False, highlightDirectAnnotations=False, highlightIntersection=True)\n",
    "display(IPython.core.display.SVG(commonGraph.draw(prog='dot',format='svg')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"626pt\" height=\"599pt\" viewBox=\"0.00 0.00 626.00 599.00\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 595)\">\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-595 622,-595 622,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_tools</title>\n",
       "</g>\n",
       "<!-- qiime2 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>qiime2</title>\n",
       "<ellipse fill=\"#ffffff\" stroke=\"blue\" cx=\"305.5\" cy=\"-34\" rx=\"45.21\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"305.5\" y=\"-28.95\" font-family=\"Times,serif\" font-size=\"14.00\">QIIME 2</text>\n",
       "</g>\n",
       "<!-- topic_3941 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>topic_3941</title>\n",
       "<polygon fill=\"#fefffe\" stroke=\"black\" points=\"376.75,-129 248.25,-129 248.25,-88 376.75,-88 376.75,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"312.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Metatranscriptomics</text>\n",
       "<text text-anchor=\"middle\" x=\"312.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3941)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3941 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3941</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M307.16,-52.18C307.86,-59.49 308.71,-68.25 309.52,-76.62\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"310.46,-86.41 305.02,-76.88 310.1,-82.64 309.5,-76.45 309.5,-76.45 309.5,-76.45 310.1,-82.64 313.98,-76.02 310.46,-86.41\"/>\n",
       "</g>\n",
       "<!-- topic_3697 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>topic_3697</title>\n",
       "<polygon fill=\"#fafffa\" stroke=\"black\" points=\"548.75,-129 432.25,-129 432.25,-88 548.75,-88 548.75,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"490.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Microbial ecology</text>\n",
       "<text text-anchor=\"middle\" x=\"490.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3697)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3697 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3697</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M336.88,-47.3C362.17,-57.21 398.48,-71.44 429.45,-83.58\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"438.59,-87.16 427.64,-87.7 435.07,-85.78 429.28,-83.51 429.28,-83.51 429.28,-83.51 435.07,-85.78 430.93,-79.32 438.59,-87.16\"/>\n",
       "</g>\n",
       "<!-- topic_3174 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>topic_3174</title>\n",
       "<polygon fill=\"#f7fff7\" stroke=\"red\" stroke-width=\"3\" points=\"230,-129 133,-129 133,-88 230,-88 230,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Metagenomics</text>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_3174)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_3174 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_3174</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M280.7,-49.5C265.25,-58.53 244.9,-70.43 226.65,-81.1\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"218.11,-86.1 224.47,-77.16 221.37,-84.19 226.74,-81.05 226.74,-81.05 226.74,-81.05 221.37,-84.19 229.01,-84.93 218.11,-86.1\"/>\n",
       "</g>\n",
       "<!-- topic_0084 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>topic_0084</title>\n",
       "<polygon fill=\"#f4fff4\" stroke=\"black\" points=\"446.38,-360 360.62,-360 360.62,-319 446.38,-319 446.38,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Phylogeny</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_0084)</text>\n",
       "</g>\n",
       "<!-- qiime2&#45;&gt;topic_0084 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>qiime2-&gt;topic_0084</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M336.62,-47.29C354.12,-55.92 374.58,-69.3 385.5,-88 426.22,-157.71 417.45,-257.62 409.48,-307.58\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"407.81,-317.39 405.06,-306.78 408.45,-313.66 409.49,-307.53 409.49,-307.53 409.49,-307.53 408.45,-313.66 413.93,-308.29 407.81,-317.39\"/>\n",
       "</g>\n",
       "<!-- vsearch -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>vsearch</title>\n",
       "<ellipse fill=\"#ffffff\" stroke=\"blue\" cx=\"188.5\" cy=\"-34\" rx=\"53.91\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"188.5\" y=\"-28.95\" font-family=\"Times,serif\" font-size=\"14.00\">VSEARCH</text>\n",
       "</g>\n",
       "<!-- vsearch&#45;&gt;topic_3174 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>vsearch-&gt;topic_3174</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M186.84,-52.18C186.17,-59.18 185.36,-67.51 184.58,-75.55\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"183.64,-85.35 180.12,-74.96 184,-81.59 184.6,-75.4 184.6,-75.4 184.6,-75.4 184,-81.59 189.08,-75.83 183.64,-85.35\"/>\n",
       "</g>\n",
       "<!-- topic_0080 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>topic_0080</title>\n",
       "<polygon fill=\"#caffca\" stroke=\"black\" points=\"115,-129 0,-129 0,-88 115,-88 115,-129\"/>\n",
       "<text text-anchor=\"middle\" x=\"57.5\" y=\"-111.7\" font-family=\"Times,serif\" font-size=\"14.00\">Sequence analysis</text>\n",
       "<text text-anchor=\"middle\" x=\"57.5\" y=\"-95.2\" font-family=\"Times,serif\" font-size=\"14.00\">(topic_0080)</text>\n",
       "</g>\n",
       "<!-- vsearch&#45;&gt;topic_0080 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>vsearch-&gt;topic_0080</title>\n",
       "<path fill=\"none\" stroke=\"blue\" stroke-dasharray=\"5,2\" d=\"M161.36,-50.02C144.56,-59.32 122.57,-71.49 103.14,-82.24\"/>\n",
       "<polygon fill=\"blue\" stroke=\"blue\" points=\"94.63,-86.95 101.2,-78.17 97.93,-85.12 103.37,-82.11 103.37,-82.11 103.37,-82.11 97.93,-85.12 105.55,-86.05 94.63,-86.95\"/>\n",
       "</g>\n",
       "<!-- topic_0203 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>topic_0203</title>\n",
       "<polygon fill=\"#ccffcc\" stroke=\"black\" points=\"346.5,-283 240.5,-283 240.5,-242 346.5,-242 346.5,-283\"/>\n",
       "<text text-anchor=\"middle\" x=\"293.5\" y=\"-265.7\" font-family=\"Times,serif\" font-size=\"14.00\">Gene expression</text>\n",
       "<text text-anchor=\"middle\" x=\"293.5\" y=\"-249.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0203</text>\n",
       "</g>\n",
       "<!-- topic_3941&#45;&gt;topic_0203 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>topic_3941-&gt;topic_0203</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M318.47,-129.36C323.48,-149.07 328.89,-179.99 322.5,-206 320.36,-214.74 316.57,-223.62 312.42,-231.65\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"309.41,-229.86 307.61,-240.3 315.53,-233.26 309.41,-229.86\"/>\n",
       "</g>\n",
       "<!-- topic_3308 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>topic_3308</title>\n",
       "<polygon fill=\"#f1fff1\" stroke=\"black\" points=\"313.75,-206 209.25,-206 209.25,-165 313.75,-165 313.75,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"261.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Transcriptomics</text>\n",
       "<text text-anchor=\"middle\" x=\"261.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3308</text>\n",
       "</g>\n",
       "<!-- topic_3941&#45;&gt;topic_3308 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>topic_3941-&gt;topic_3308</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M299.1,-129.21C293.7,-137.15 287.35,-146.49 281.42,-155.2\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"278.57,-153.18 275.84,-163.41 284.36,-157.11 278.57,-153.18\"/>\n",
       "</g>\n",
       "<!-- topic_0610 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>topic_0610</title>\n",
       "<polygon fill=\"#f2fff2\" stroke=\"red\" stroke-width=\"3\" points=\"508.88,-206 432.12,-206 432.12,-165 508.88,-165 508.88,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"470.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Ecology</text>\n",
       "<text text-anchor=\"middle\" x=\"470.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0610</text>\n",
       "</g>\n",
       "<!-- topic_3697&#45;&gt;topic_0610 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>topic_3697-&gt;topic_0610</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M485.25,-129.21C483.33,-136.4 481.1,-144.74 478.97,-152.72\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"475.65,-151.59 476.46,-162.15 482.42,-153.39 475.65,-151.59\"/>\n",
       "</g>\n",
       "<!-- topic_3301 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>topic_3301</title>\n",
       "<polygon fill=\"#f8fff8\" stroke=\"black\" points=\"618,-206 527,-206 527,-165 618,-165 618,-206\"/>\n",
       "<text text-anchor=\"middle\" x=\"572.5\" y=\"-188.7\" font-family=\"Times,serif\" font-size=\"14.00\">Microbiology</text>\n",
       "<text text-anchor=\"middle\" x=\"572.5\" y=\"-172.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3301</text>\n",
       "</g>\n",
       "<!-- topic_3697&#45;&gt;topic_3301 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>topic_3697-&gt;topic_3301</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M512.04,-129.21C521.39,-137.75 532.48,-147.9 542.6,-157.16\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"539.96,-159.48 549.7,-163.64 544.68,-154.31 539.96,-159.48\"/>\n",
       "</g>\n",
       "<!-- topic_3174&#45;&gt;topic_0610 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>topic_3174-&gt;topic_0610</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M231.4,-126.59C234.13,-127.43 236.85,-128.24 239.5,-129 301.02,-146.69 372.85,-163.33 419.88,-173.69\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"418.99,-177.08 429.51,-175.8 420.49,-170.24 418.99,-177.08\"/>\n",
       "</g>\n",
       "<!-- topic_0622 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>topic_0622</title>\n",
       "<polygon fill=\"#d5ffd5\" stroke=\"red\" stroke-width=\"3\" points=\"222.88,-283 146.12,-283 146.12,-242 222.88,-242 222.88,-283\"/>\n",
       "<text text-anchor=\"middle\" x=\"184.5\" y=\"-265.7\" font-family=\"Times,serif\" font-size=\"14.00\">Genomics</text>\n",
       "<text text-anchor=\"middle\" x=\"184.5\" y=\"-249.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0622</text>\n",
       "</g>\n",
       "<!-- topic_3174&#45;&gt;topic_0622 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>topic_3174-&gt;topic_0622</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M181.91,-130.32C182.42,-155.9 183.28,-199.58 183.87,-229.46\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"180.37,-229.49 184.06,-239.42 187.37,-229.35 180.37,-229.49\"/>\n",
       "</g>\n",
       "<!-- topic_3307 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>topic_3307</title>\n",
       "<polygon fill=\"#45ff45\" stroke=\"red\" stroke-width=\"3\" points=\"225.5,-437 83.5,-437 83.5,-396 225.5,-396 225.5,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Computational biology</text>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3307</text>\n",
       "</g>\n",
       "<!-- topic_0084&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>topic_0084-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M360.21,-356.73C356.93,-357.87 353.67,-358.97 350.5,-360 313.62,-371.99 272.37,-383.89 237.26,-393.58\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"236.75,-390.09 228.04,-396.11 238.61,-396.84 236.75,-390.09\"/>\n",
       "</g>\n",
       "<!-- topic_3299 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>topic_3299</title>\n",
       "<polygon fill=\"#efffef\" stroke=\"black\" points=\"469.25,-437 337.75,-437 337.75,-396 469.25,-396 469.25,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Evolutionary biology</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3299</text>\n",
       "</g>\n",
       "<!-- topic_0084&#45;&gt;topic_3299 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>topic_0084-&gt;topic_3299</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M403.5,-360.21C403.5,-367.55 403.5,-376.09 403.5,-384.22\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"400,-384.15 403.5,-394.15 407,-384.15 400,-384.15\"/>\n",
       "</g>\n",
       "<!-- topic_3321 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>topic_3321</title>\n",
       "<polygon fill=\"#a6ffa6\" stroke=\"black\" points=\"341.62,-360 221.38,-360 221.38,-319 341.62,-319 341.62,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Molecular genetics</text>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3321</text>\n",
       "</g>\n",
       "<!-- topic_0203&#45;&gt;topic_3321 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>topic_0203-&gt;topic_3321</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M290.35,-283.21C289.16,-290.64 287.77,-299.29 286.46,-307.51\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"283.04,-306.74 284.91,-317.17 289.95,-307.85 283.04,-306.74\"/>\n",
       "</g>\n",
       "<!-- topic_3308&#45;&gt;topic_0203 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>topic_3308-&gt;topic_0203</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M269.91,-206.21C273.19,-213.9 277.02,-222.89 280.64,-231.35\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"277.3,-232.45 284.44,-240.27 283.74,-229.7 277.3,-232.45\"/>\n",
       "</g>\n",
       "<!-- topic_3308&#45;&gt;topic_0622 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>topic_3308-&gt;topic_0622</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M241.27,-206.21C232.95,-214.31 223.14,-223.87 214.04,-232.73\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"211.75,-230.07 207.03,-239.56 216.63,-235.09 211.75,-230.07\"/>\n",
       "</g>\n",
       "<!-- topic_3070 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>topic_3070</title>\n",
       "<polygon fill=\"#61ff61\" stroke=\"red\" stroke-width=\"3\" points=\"441.88,-514 365.12,-514 365.12,-473 441.88,-473 441.88,-514\"/>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-496.7\" font-family=\"Times,serif\" font-size=\"14.00\">Biology</text>\n",
       "<text text-anchor=\"middle\" x=\"403.5\" y=\"-480.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3070</text>\n",
       "</g>\n",
       "<!-- topic_0610&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>topic_0610-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M477.52,-207.4C491.28,-252.29 517.28,-360.07 478.5,-437 472.58,-448.75 462.87,-458.63 452.37,-466.64\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"450.42,-463.73 444.25,-472.33 454.44,-469.46 450.42,-463.73\"/>\n",
       "</g>\n",
       "<!-- topic_3301&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>topic_3301-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M571.62,-206.3C568.65,-251.79 555.69,-365.36 499.5,-437 487.5,-452.3 470.02,-464.17 453.25,-472.93\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"451.75,-469.77 444.32,-477.32 454.84,-476.05 451.75,-469.77\"/>\n",
       "</g>\n",
       "<!-- topic_3391 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>topic_3391</title>\n",
       "<polygon fill=\"#c2ffc2\" stroke=\"red\" stroke-width=\"3\" points=\"90.88,-360 14.12,-360 14.12,-319 90.88,-319 90.88,-360\"/>\n",
       "<text text-anchor=\"middle\" x=\"52.5\" y=\"-342.7\" font-family=\"Times,serif\" font-size=\"14.00\">Omics</text>\n",
       "<text text-anchor=\"middle\" x=\"52.5\" y=\"-326.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3391</text>\n",
       "</g>\n",
       "<!-- topic_0622&#45;&gt;topic_3391 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>topic_0622-&gt;topic_3391</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M147.73,-284.39C132.57,-293.01 114.83,-303.09 98.8,-312.2\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"97.35,-308.99 90.39,-316.97 100.81,-315.08 97.35,-308.99\"/>\n",
       "</g>\n",
       "<!-- topic_0003 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>topic_0003</title>\n",
       "<polygon fill=\"#00ff00\" stroke=\"red\" stroke-width=\"3\" points=\"192.88,-591 116.12,-591 116.12,-550 192.88,-550 192.88,-591\"/>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-573.7\" font-family=\"Times,serif\" font-size=\"14.00\">Topic</text>\n",
       "<text text-anchor=\"middle\" x=\"154.5\" y=\"-557.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_0003</text>\n",
       "</g>\n",
       "<!-- topic_3307&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>topic_3307-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M154.5,-438.32C154.5,-463.9 154.5,-507.58 154.5,-537.46\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"151,-537.42 154.5,-547.42 158,-537.42 151,-537.42\"/>\n",
       "</g>\n",
       "<!-- topic_3299&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>topic_3299-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M403.5,-437.21C403.5,-444.24 403.5,-452.36 403.5,-460.17\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"400,-460.1 403.5,-470.1 407,-460.1 400,-460.1\"/>\n",
       "</g>\n",
       "<!-- topic_3321&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>topic_3321-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M247.8,-360.4C232.95,-369.17 215.29,-379.6 199.35,-389.01\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"197.82,-385.85 190.99,-393.95 201.38,-391.88 197.82,-385.85\"/>\n",
       "</g>\n",
       "<!-- topic_3053 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>topic_3053</title>\n",
       "<polygon fill=\"#92ff92\" stroke=\"black\" points=\"319.88,-437 243.12,-437 243.12,-396 319.88,-396 319.88,-437\"/>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-419.7\" font-family=\"Times,serif\" font-size=\"14.00\">Genetics</text>\n",
       "<text text-anchor=\"middle\" x=\"281.5\" y=\"-403.2\" font-family=\"Times,serif\" font-size=\"14.00\">topic_3053</text>\n",
       "</g>\n",
       "<!-- topic_3321&#45;&gt;topic_3053 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>topic_3321-&gt;topic_3053</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M281.5,-360.21C281.5,-367.55 281.5,-376.09 281.5,-384.22\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"278,-384.15 281.5,-394.15 285,-384.15 278,-384.15\"/>\n",
       "</g>\n",
       "<!-- topic_3053&#45;&gt;topic_3070 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>topic_3053-&gt;topic_3070</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M313.87,-437.4C328,-446.09 344.78,-456.4 359.98,-465.75\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"358.05,-468.67 368.41,-470.93 361.72,-462.71 358.05,-468.67\"/>\n",
       "</g>\n",
       "<!-- topic_3070&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>topic_3070-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M363.77,-506.47C320.68,-519.45 251.79,-540.2 204.91,-554.32\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"204.1,-550.9 195.54,-557.14 206.12,-557.61 204.1,-550.9\"/>\n",
       "</g>\n",
       "<!-- topic_3391&#45;&gt;topic_0003 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>topic_3391-&gt;topic_0003</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M55.29,-361.41C58.41,-381.36 64.4,-411.98 74.5,-437 89.57,-474.34 114.41,-513.37 132.42,-539.31\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"129.46,-541.19 138.08,-547.35 135.18,-537.15 129.46,-541.19\"/>\n",
       "</g>\n",
       "<!-- topic_0080&#45;&gt;topic_3307 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>topic_0080-&gt;topic_3307</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M63.72,-129.11C80.09,-180.77 124.53,-320.96 144.52,-384.02\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"141.14,-384.94 147.5,-393.41 147.81,-382.82 141.14,-384.94\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# color by the number of tools\n",
    "dfTopicNbToolsTransitive = dfToolTopicTransitive[['tool', 'topic']].groupby(by=['topic'], as_index=False).size()\n",
    "dictTopicNbToolsTransitive = {}\n",
    "for (conceptURI, conceptNbTools) in dfTopicNbToolsTransitive[['topic', 'size']].set_index('topic').to_dict()['size'].items():\n",
    "    dictTopicNbToolsTransitive[conceptURI.replace(edamURI, \"\")] = conceptNbTools\n",
    "\n",
    "dfOperationNbToolsTransitive = dfToolOperationTransitive[['tool', 'operation']].groupby(by=['operation'], as_index=False).size()\n",
    "dictOperationNbToolsTransitive = {}\n",
    "for (conceptURI, conceptNbTools) in dfOperationNbToolsTransitive[['operation', 'size']].set_index('operation').to_dict()['size'].items():\n",
    "    dictOperationNbToolsTransitive[conceptURI.replace(edamURI, \"\")] = conceptNbTools\n",
    "    \n",
    "dictTopicNbOccurrences = collections.defaultdict(lambda: 0)\n",
    "dictOperationsNbOccurrences = collections.defaultdict(lambda: 0)\n",
    "\n",
    "for currentTool in tools:\n",
    "    for currentTopic in [x.replace(edamURI, \"\") for x in dfToolTopicTransitive[dfToolTopicTransitive['tool'] == biotoolsURI+currentTool]['topic'].to_list()]:\n",
    "        dictTopicNbOccurrences[currentTopic] += 1\n",
    "    for currentOperation in [x.replace(edamURI, \"\") for x in dfToolOperationTransitive[dfToolOperationTransitive['tool'] == biotoolsURI+currentTool]['operation'].to_list()]:\n",
    "        dictOperationsNbOccurrences[currentOperation] += 1\n",
    "\n",
    "# color by the number of tools\n",
    "colorGraphNodesAccordingToScore(commonGraph, dictTopicNbToolsTransitive, dictOperationNbToolsTransitive, color=\"green\")\n",
    "display(IPython.core.display.SVG(commonGraph.draw(prog='dot',format='svg')))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biotoolsEDAMvisualization",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
