{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figures Assessing bioinformatics software annotations: bio.tools case-study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 1 : Direct and inferred EDAM annotations for the scientific topic of Qiime2 and Vsearch, two metagenomics tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Python 3.13.1\n",
    "#Import library \n",
    "import pandas\n",
    "import IPython\n",
    "import collections\n",
    "import pygraphviz as pgv\n",
    "#Read csv \n",
    "dfTool = pandas.read_csv(\"results/dfTool.tsv.bz2\", sep=\"\\t\") #all Tool and toolLabel\n",
    "dfToolTopicTransitive = pandas.read_csv(\"results/dfToolTopicTransitive.tsv.bz2\", sep=\"\\t\")#Tool, topic, topicLabel transitive\n",
    "dfToolOperationTransitive = pandas.read_csv(\"results/dfToolOperationTransitive.tsv.bz2\", sep=\"\\t\")# tool, operation, operationLabel transitive\n",
    "#Import prefix : \n",
    "prefixes = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>\n",
    "\n",
    "PREFIX bt: <https://bio.tools/>\n",
    "PREFIX biotools: <https://bio.tools/ontology/>\n",
    "PREFIX bsc: <http://bioschemas.org/>\n",
    "PREFIX bsct: <http://bioschemas.org/types/>\n",
    "PREFIX edam: <http://edamontology.org/>\n",
    "PREFIX sc: <http://schema.org/>\n",
    "PREFIX schema: <https://schema.org/>\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#Link tool\n",
    "biotoolsURI = \"https://bio.tools/\"\n",
    "biotoolsOntologyURI = \"https://bio.tools/ontology/\"\n",
    "edamURI = \"http://edamontology.org/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>operation</th>\n",
       "      <th>operationLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2428</td>\n",
       "      <td>Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0004</td>\n",
       "      <td>Operation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0484</td>\n",
       "      <td>SNP detection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_3227</td>\n",
       "      <td>Variant calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2478</td>\n",
       "      <td>Nucleic acid sequence analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249653</th>\n",
       "      <td>https://bio.tools/gbc</td>\n",
       "      <td>http://edamontology.org/operation_2480</td>\n",
       "      <td>Structure analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249654</th>\n",
       "      <td>https://bio.tools/gbc</td>\n",
       "      <td>http://edamontology.org/operation_0244</td>\n",
       "      <td>Simulation analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249655</th>\n",
       "      <td>https://bio.tools/gbc</td>\n",
       "      <td>http://edamontology.org/operation_0250</td>\n",
       "      <td>Protein property calculation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249656</th>\n",
       "      <td>https://bio.tools/gbc</td>\n",
       "      <td>http://edamontology.org/operation_3438</td>\n",
       "      <td>Calculation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249657</th>\n",
       "      <td>https://bio.tools/gbc</td>\n",
       "      <td>http://edamontology.org/operation_2423</td>\n",
       "      <td>Prediction and recognition</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249658 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            tool                               operation  \\\n",
       "0       https://bio.tools/pmirkb  http://edamontology.org/operation_2428   \n",
       "1       https://bio.tools/pmirkb  http://edamontology.org/operation_0004   \n",
       "2       https://bio.tools/pmirkb  http://edamontology.org/operation_0484   \n",
       "3       https://bio.tools/pmirkb  http://edamontology.org/operation_3227   \n",
       "4       https://bio.tools/pmirkb  http://edamontology.org/operation_2478   \n",
       "...                          ...                                     ...   \n",
       "249653     https://bio.tools/gbc  http://edamontology.org/operation_2480   \n",
       "249654     https://bio.tools/gbc  http://edamontology.org/operation_0244   \n",
       "249655     https://bio.tools/gbc  http://edamontology.org/operation_0250   \n",
       "249656     https://bio.tools/gbc  http://edamontology.org/operation_3438   \n",
       "249657     https://bio.tools/gbc  http://edamontology.org/operation_2423   \n",
       "\n",
       "                        operationLabel  \n",
       "0                           Validation  \n",
       "1                            Operation  \n",
       "2                        SNP detection  \n",
       "3                      Variant calling  \n",
       "4       Nucleic acid sequence analysis  \n",
       "...                                ...  \n",
       "249653              Structure analysis  \n",
       "249654             Simulation analysis  \n",
       "249655    Protein property calculation  \n",
       "249656                     Calculation  \n",
       "249657      Prediction and recognition  \n",
       "\n",
       "[249658 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfToolOperationTransitive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tool_url(tool_name: str) -> str:\n",
    "    \"\"\"Returns the full bio.tools URL for a given tool name, ensuring the correct format.\"\"\"\n",
    "    if tool_name.startswith(\"https://bio.tools/\"):\n",
    "        return tool_name\n",
    "    return f\"https://bio.tools/{tool_name}\"\n",
    "\n",
    "\n",
    "def getToolTopics(tool):\n",
    "    \"\"\"\n",
    "    Returns a list of (topic, topicLabel) tuples for the given tool.\n",
    "    \"\"\"\n",
    "    filtered = dfToolTopicTransitive[dfToolTopicTransitive['tool'] == tool]\n",
    "    return list(filtered[['topic', 'topicLabel']].itertuples(index=False, name=None))\n",
    "\n",
    "\n",
    "def getToolOperations(tool):\n",
    "    \"\"\"\n",
    "    Returns a list of (operation, operationLabel) tuples for the given tool.\n",
    "    \"\"\"\n",
    "    filtered = dfToolOperationTransitive[dfToolOperationTransitive['tool'] == tool]\n",
    "    return list(filtered[['operation', 'operationLabel']].itertuples(index=False, name=None))\n",
    "\n",
    "\n",
    "def getToolsCommonTopics(listToolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the topics associated to all the tools of a list.\n",
    "\n",
    "    Keyword arguments:\n",
    "    listToolURI -- list of the URIs for the tools\n",
    "    transitive -- also consider the ancestors of the topics directly associated to a tool (default: False)\n",
    "    \"\"\"\n",
    "    commonConcepts = []\n",
    "    if len(listToolURI) > 0:\n",
    "        commonConcepts = set(getToolTopics(listToolURI[0], transitive=transitive))\n",
    "    for toolURI in listToolURI[1:]:\n",
    "        currentConcepts = set(getToolTopics(toolURI, transitive=transitive))\n",
    "        commonConcepts = commonConcepts.intersection(currentConcepts)\n",
    "    return list(commonConcepts)\n",
    "\n",
    "\n",
    "def getToolsCommonOperations(listToolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the operations associated to all the tools of a list.\n",
    "\n",
    "    Keyword arguments:\n",
    "    listToolURI -- list of the URIs for the tools\n",
    "    transitive -- also consider the ancestors of the operations directly associated to a tool (default: False)\n",
    "    \"\"\"\n",
    "    commonConcepts = []\n",
    "    if len(listToolURI) > 0:\n",
    "        commonConcepts = set(getToolOperations(listToolURI[0], transitive=transitive))\n",
    "    for toolURI in listToolURI[1:]:\n",
    "        currentConcepts = set(getToolOperations(toolURI, transitive=transitive))\n",
    "        commonConcepts = commonConcepts.intersection(currentConcepts)\n",
    "    return list(commonConcepts)\n",
    "\n",
    "\n",
    "def colorGraphNodesAccordingToScore(graph, dictTopicScore, dictOperationScore, color=\"red\"):\n",
    "    \"\"\"Modify nodes color according to a score associated to the node. Nodes that are not associated to a score are unaffected.\n",
    "\n",
    "    Keyword arguments:\n",
    "    graph -- the graph to be modified\n",
    "    dictTopicScore -- a dictionary {nodeIdent -> score} for topic nodes\n",
    "    dictOperationScore -- a dictionary {nodeIdent -> score} for operation nodes\n",
    "    color -- name of the color that will vary according to scores. Possible values are \"red\", \"green\" or \"blue\" (default: \"red\")\n",
    "    \"\"\"\n",
    "    topicMaxValue = max(dictTopicScore.values(), default=0)\n",
    "    operationMaxValue = max(dictOperationScore.values(), default=0)\n",
    "    for currentNode in graph.nodes_iter():\n",
    "        if \"nodeType\" in currentNode.attr.keys():\n",
    "            currentNodeType = currentNode.attr['nodeType']\n",
    "            #currentNode.attr['style'] = conceptStyle[currentNodeType]\n",
    "            if currentNodeType == \"Topic\":\n",
    "                if currentNode in dictTopicScore.keys():\n",
    "                    graph.get_node(currentNode).attr['fillcolor'] = getScoreColorRGB(dictTopicScore[currentNode], topicMaxValue, color=color)\n",
    "            elif currentNodeType == \"Operation\":\n",
    "                if currentNode in dictOperationScore.keys():\n",
    "                    graph.get_node(currentNode).attr['fillcolor'] = getScoreColorRGB(dictOperationScore[currentNode], operationMaxValue, color=color)\n",
    "\n",
    "\n",
    "def addToolsAndAnnotationsToGraph(listToolURI, graph=None, showTopics=True, showOperations=True, highlightDirectAnnotations=False, highlightIntersection=False):\n",
    "    \"\"\"Return a graph representing tools and their EDAM annotations.\n",
    "\n",
    "    Keyword arguments:\n",
    "    listToolURI -- list of the URIs for the tools\n",
    "    graph -- the graph in which the tool and its annotations are added. A new graph is created if the value is None. (default: None)\n",
    "    showTopics -- should the topics annotating the tool be considered (default: True)\n",
    "    showOperations -- should the operations annotating the tool be considered (default: True)\n",
    "    highlightDirectAnnotations -- should the topics or operations annotated directly be highligthed (default:False)\n",
    "    highlightIntersection -- should the common topics or operation common to all the tools be highlighted\n",
    "    \"\"\"\n",
    "    if graph is None:\n",
    "        graph = pgv.AGraph(directed=True, rankdir=\"BT\")\n",
    "    for toolURI in listToolURI:\n",
    "        addToolAndAnnotationsToGraph(toolURI, graph=graph, showTopics=showTopics, showOperations=showOperations, highlightDirectAnnotations=highlightDirectAnnotations)\n",
    "    if highlightIntersection:\n",
    "        commonConcepts = []\n",
    "        if showTopics:\n",
    "            commonConcepts += getToolsCommonTopics(listToolURI, transitive=True)\n",
    "        if showOperations:\n",
    "            commonConcepts += getToolsCommonOperations(listToolURI, transitive=True)\n",
    "        for (currentConcept, currentLabel) in commonConcepts:\n",
    "            currentConceptIdent = currentConcept.replace(edamURI, \"\")\n",
    "            graph.get_node(currentConceptIdent).attr['color'] = \"red\"\n",
    "    return graph\n",
    "\n",
    "\n",
    "def getScoreColorRGB(scoreValue, scoreMaxValue, color=\"red\"):\n",
    "    \"\"\"Return the RGB color (in hex) associated to a score, varying from white (0 score)\n",
    "    to a target color at the maximum score.\n",
    "\n",
    "    Keyword arguments:\n",
    "    scoreValue -- the score to consider (positive number)\n",
    "    scoreMaxValue -- the maximum value for the score (positive number, >= scoreValue)\n",
    "    color -- name of the color for the gradient. Possible values are:\n",
    "             \"red\", \"green\", \"blue\", \"orange\", \"yellow\", \"pink\", \"grey\".\n",
    "             (default: \"red\")\n",
    "    \"\"\"\n",
    "    # Avoid division by 0.\n",
    "    scoreMaxValue = max(1, scoreMaxValue)\n",
    "    fraction = scoreValue / scoreMaxValue\n",
    "\n",
    "    # Define the target color values.\n",
    "    if color == \"red\":\n",
    "        target = (255, 0, 0)\n",
    "    elif color == \"green\":\n",
    "        target = (0, 255, 0)\n",
    "    elif color == \"blue\":\n",
    "        target = (0, 0, 255)\n",
    "    elif color == \"orange\":\n",
    "        target = (255, 165, 0)\n",
    "    elif color == \"yellow\":\n",
    "        target = (255, 255, 0)\n",
    "    elif color == \"pink\":\n",
    "        target = (255, 192, 203)  # Light pink (alternative: hot pink (255,105,180))\n",
    "    elif color == \"grey\":\n",
    "        target = (128, 128, 128)\n",
    "    else:\n",
    "        return \"#ffffff\"\n",
    "\n",
    "    # Interpolate from white (255,255,255) to the target color.\n",
    "    r = int(255 - fraction * (255 - target[0]))\n",
    "    g = int(255 - fraction * (255 - target[1]))\n",
    "    b = int(255 - fraction * (255 - target[2]))\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(r, g, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb tools: 29993\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'addToolAndAnnotationsToGraph' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 9\u001b[0m\n\u001b[1;32m      6\u001b[0m fileFormat \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpng\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      8\u001b[0m listToolURI \u001b[38;5;241m=\u001b[39m [biotoolsURI\u001b[38;5;241m+\u001b[39mx \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m tools]\n\u001b[0;32m----> 9\u001b[0m commonGraph \u001b[38;5;241m=\u001b[39m \u001b[43maddToolsAndAnnotationsToGraph\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlistToolURI\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgraph\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshowTopics\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshowOperations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhighlightDirectAnnotations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhighlightIntersection\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m display(IPython\u001b[38;5;241m.\u001b[39mcore\u001b[38;5;241m.\u001b[39mdisplay\u001b[38;5;241m.\u001b[39mSVG(commonGraph\u001b[38;5;241m.\u001b[39mdraw(prog\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdot\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;28mformat\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m)))\n\u001b[1;32m     11\u001b[0m fileName \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28msorted\u001b[39m(tools)) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_operations\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "Cell \u001b[0;32mIn[13], line 93\u001b[0m, in \u001b[0;36maddToolsAndAnnotationsToGraph\u001b[0;34m(listToolURI, graph, showTopics, showOperations, highlightDirectAnnotations, highlightIntersection)\u001b[0m\n\u001b[1;32m     91\u001b[0m     graph \u001b[38;5;241m=\u001b[39m pgv\u001b[38;5;241m.\u001b[39mAGraph(directed\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, rankdir\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBT\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m toolURI \u001b[38;5;129;01min\u001b[39;00m listToolURI:\n\u001b[0;32m---> 93\u001b[0m     \u001b[43maddToolAndAnnotationsToGraph\u001b[49m(toolURI, graph\u001b[38;5;241m=\u001b[39mgraph, showTopics\u001b[38;5;241m=\u001b[39mshowTopics, showOperations\u001b[38;5;241m=\u001b[39mshowOperations, highlightDirectAnnotations\u001b[38;5;241m=\u001b[39mhighlightDirectAnnotations)\n\u001b[1;32m     94\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m highlightIntersection:\n\u001b[1;32m     95\u001b[0m     commonConcepts \u001b[38;5;241m=\u001b[39m []\n",
      "\u001b[0;31mNameError\u001b[0m: name 'addToolAndAnnotationsToGraph' is not defined"
     ]
    }
   ],
   "source": [
    "tools = [\"qiime2\",\"vsearch\"]\n",
    "nbTools = len(dfTool)\n",
    "print(\"Nb tools: {}\".format(nbTools))\n",
    "\"-\".join(sorted(tools)) + \"_topics_operations\"\n",
    "\n",
    "fileFormat = \"png\"\n",
    "\n",
    "listToolURI = [biotoolsURI+x for x in tools]\n",
    "commonGraph = addToolsAndAnnotationsToGraph(listToolURI, graph=None, showTopics=True, showOperations=False, highlightDirectAnnotations=False, highlightIntersection=True)\n",
    "display(IPython.core.display.SVG(commonGraph.draw(prog='dot',format='svg')))\n",
    "fileName = \"-\".join(sorted(tools)) + \"_operations\"\n",
    "commonGraph.draw(path='results/{}.{}'.format(fileName, fileFormat), prog='dot',format=fileFormat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'commonGraph' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 10\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m currentOperation \u001b[38;5;129;01min\u001b[39;00m [x\u001b[38;5;241m.\u001b[39mreplace(edamURI, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m dfToolOperationTransitive[dfToolOperationTransitive[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtool\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m biotoolsURI\u001b[38;5;241m+\u001b[39mcurrentTool][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124moperation\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mto_list()]:\n\u001b[1;32m      8\u001b[0m         dictOperationsNbOccurrences[currentOperation] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m---> 10\u001b[0m colorGraphNodesAccordingToScore(\u001b[43mcommonGraph\u001b[49m, dictTopicNbOccurrences, dictOperationsNbOccurrences, color\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124morange\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     11\u001b[0m display(IPython\u001b[38;5;241m.\u001b[39mcore\u001b[38;5;241m.\u001b[39mdisplay\u001b[38;5;241m.\u001b[39mSVG(commonGraph\u001b[38;5;241m.\u001b[39mdraw(prog\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdot\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;28mformat\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m)))\n\u001b[1;32m     12\u001b[0m fileName \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28msorted\u001b[39m(tools)) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMetagenomics_topics_nbannotations_colors\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'commonGraph' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# color by the number of tools\n",
    "dictTopicNbOccurrences = collections.defaultdict(lambda: 0)\n",
    "dictOperationsNbOccurrences = collections.defaultdict(lambda: 0)\n",
    "for currentTool in tools:\n",
    "    for currentTopic in [x.replace(edamURI, \"\") for x in dfToolTopicTransitive[dfToolTopicTransitive['tool'] == biotoolsURI+currentTool]['topic'].to_list()]:\n",
    "        dictTopicNbOccurrences[currentTopic] += 1\n",
    "    for currentOperation in [x.replace(edamURI, \"\") for x in dfToolOperationTransitive[dfToolOperationTransitive['tool'] == biotoolsURI+currentTool]['operation'].to_list()]:\n",
    "        dictOperationsNbOccurrences[currentOperation] += 1\n",
    "\n",
    "colorGraphNodesAccordingToScore(commonGraph, dictTopicNbOccurrences, dictOperationsNbOccurrences, color=\"orange\")\n",
    "display(IPython.core.display.SVG(commonGraph.draw(prog='dot',format='svg')))\n",
    "fileName = \"-\".join(sorted(tools)) + \"Metagenomics_topics_nbannotations_colors\"\n",
    "commonGraph.draw(path='results/{}.{}'.format(fileName, fileFormat), prog='dot',format=fileFormat)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biotoolsEDAMvisualization",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
