{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec3f3c82",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import library \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import IPython\n",
    "import sparqldataframe\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "628ca03a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Configuration SPARQL end point \n",
    "endpointURL = \"http://localhost:3030/biotoolsEdam/query\"\n",
    "rdfFormat = \"turtle\"\n",
    "\n",
    "#Import prefix : \n",
    "prefixes = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>\n",
    "\n",
    "PREFIX bt: <https://bio.tools/>\n",
    "PREFIX biotools: <https://bio.tools/ontology/>\n",
    "PREFIX bsc: <http://bioschemas.org/>\n",
    "PREFIX bsct: <http://bioschemas.org/types/>\n",
    "PREFIX edam: <http://edamontology.org/>\n",
    "PREFIX sc: <http://schema.org/>\n",
    "PREFIX schema: <https://schema.org/>\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#Link tool\n",
    "biotoolsURI = \"https://bio.tools/\"\n",
    "biotoolsOntologyURI = \"https://bio.tools/ontology/\"\n",
    "edamURI = \"http://edamontology.org/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e6c733d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def displaySparqlResults(results):\n",
    "    '''\n",
    "    Displays as HTML the result of a SPARQLWrapper query in a Jupyter notebook.\n",
    "    \n",
    "        Parameters:\n",
    "            results (dictionnary): the result of a call to SPARQLWrapper.query().convert()\n",
    "    '''\n",
    "    variableNames = results['head']['vars']\n",
    "    #tableCode = '<table><tr><th>{}</th></tr><tr>{}</tr></table>'.format('</th><th>'.join(variableNames), '</tr><tr>'.join('<td>{}</td>'.format('</td><td>'.join([row[vName]['value'] for vName in variableNames]))for row in results[\"results\"][\"bindings\"]))\n",
    "    tableCode = '<table><tr><th>{}</th></tr><tr>{}</tr></table>'.format('</th><th>'.join(variableNames), '</tr><tr>'.join('<td>{}</td>'.format('</td><td>'.join([row[vName]['value'] if vName in row.keys() else \"&nbsp;\" for vName in variableNames]))for row in results[\"results\"][\"bindings\"]))\n",
    "    IPython.display.display(IPython.display.HTML(tableCode))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8bb6f0ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions for querying topics, operations and associated labels for a tool. \n",
    "def getToolLabel(toolURI):\n",
    "    \"\"\"Return the label of a tool (or empty string if no label is present).\n",
    "\n",
    "    Keyword arguments:\n",
    "    toolURI -- the URI for the tool\n",
    "    \"\"\"\n",
    "    \n",
    "    if toolURI.startswith(\"http\"):\n",
    "        toolURI = \"<\" + toolURI + \">\"\n",
    "    query = \"\"\"\n",
    "SELECT DISTINCT ?tool ?toolLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  OPTIONAL { ?tool sc:name ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?toolLabel)\n",
    "}\n",
    "\"\"\"\n",
    "    sparql = SPARQLWrapper(endpointURL)\n",
    "    sparql.setQuery(prefixes+query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.queryAndConvert()\n",
    "    return results[\"results\"][\"bindings\"][0][\"toolLabel\"][\"value\"]\n",
    "\n",
    "\n",
    "def getToolURIByLabel(toolLabel):\n",
    "    \"\"\"Return the URI of a tool designated by its label (or None).\n",
    "\n",
    "    Keyword arguments:\n",
    "    toolLabel -- the label for the tool\n",
    "    \"\"\"\n",
    "    \n",
    "    query = \"\"\"\n",
    "SELECT DISTINCT ?tool ?toolLabel\n",
    "WHERE {\n",
    "  VALUES ?toolLabel { \\\"\"\"\" + toolLabel + \"\"\"\\\" }\n",
    "\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:name ?toolLabel .\n",
    "}\n",
    "\"\"\"\n",
    "    sparql = SPARQLWrapper(endpointURL)\n",
    "    sparql.setQuery(prefixes+query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.queryAndConvert()\n",
    "    return None if len(results[\"results\"][\"bindings\"]) == 0 else results[\"results\"][\"bindings\"][0][\"tool\"][\"value\"]\n",
    "\n",
    "\n",
    "def getToolTopics(toolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the topics associated to a tool.\n",
    "\n",
    "    Keyword arguments:\n",
    "    toolURI -- the URI for the tool\n",
    "    transitive -- also consider the ancestors of the topics directly associated to the tool (default: False)\n",
    "    \"\"\"\n",
    "    \n",
    "    if toolURI.startswith(\"http\"):\n",
    "        toolURI = \"<\" + toolURI + \">\"\n",
    "    transitiveClause = \"/(rdfs:subClassOf*)\" if transitive else \"\"\n",
    "    query = \"\"\"\n",
    "SELECT DISTINCT ?tool ?topic ?topicLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:applicationSubCategory\"\"\" + transitiveClause + \"\"\" ?topic .\n",
    "  ?topic rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?topic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?topic rdfs:label ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?topicLabel)\n",
    "}\n",
    "\"\"\"\n",
    "    sparql = SPARQLWrapper(endpointURL)\n",
    "    sparql.setQuery(prefixes+query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.queryAndConvert()\n",
    "    toolTopics = [(result[\"topic\"][\"value\"], result[\"topicLabel\"][\"value\"]) for result in results[\"results\"][\"bindings\"]]\n",
    "    return toolTopics\n",
    "\n",
    "\n",
    "def getToolOperations(toolURI, transitive=False):\n",
    "    \"\"\"Return the list of the (URI, label) tuples for the operations associated to a tool.\n",
    "\n",
    "    Keyword arguments:\n",
    "    toolURI -- the URI for the tool\n",
    "    transitive -- also consider the ancestors of the operations directly associated to the tool (default: False)\n",
    "    \"\"\"\n",
    "    \n",
    "    if toolURI.startswith(\"http\"):\n",
    "        toolURI = \"<\" + toolURI + \">\"\n",
    "    transitiveClause = \"/(rdfs:subClassOf*)\" if transitive else \"\"\n",
    "    query = \"\"\"\n",
    "SELECT DISTINCT ?tool ?operation ?operationLabel\n",
    "WHERE {\n",
    "  VALUES ?tool { \"\"\" + toolURI + \"\"\" }\n",
    "\n",
    "  ?tool sc:featureList\"\"\" + transitiveClause + \"\"\" ?operation .\n",
    "  ?operation rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?operation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?operation rdfs:label ?oLabel }\n",
    "  BIND(COALESCE(?oLabel, \"\") AS ?operationLabel)\n",
    "}\n",
    "\"\"\"\n",
    "    sparql = SPARQLWrapper(endpointURL)\n",
    "    sparql.setQuery(prefixes+query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.queryAndConvert()\n",
    "    toolOperations = [(result[\"operation\"][\"value\"], result[\"operationLabel\"][\"value\"]) for result in results[\"results\"][\"bindings\"]]\n",
    "    return toolOperations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e26d2794",
   "metadata": {},
   "source": [
    "## Metrics on bio.tools : Tools with topic and operation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d875a9e",
   "metadata": {},
   "source": [
    "### Tools with SoftwareApplication class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75a62d5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number tools (SoftwareApplication) in bio.tools extraction : 30025\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# tools\n",
    "\n",
    "SELECT DISTINCT ?tool ?toolLabel\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:name ?toolLabel .\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "dfTool = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "print(f\"Number tools (SoftwareApplication) in bio.tools extraction : {len(dfTool)}\")\n",
    "nbTools = len(dfTool)\n",
    "#Can generate dfTool.tsv.bz2 : \n",
    "#dfTool.to_csv(\"../Dataframes/dfTool.tsv.bz2\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e1c6df",
   "metadata": {},
   "source": [
    "### Number of tools with at least one topic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7415701d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb tools with topic: 29616 (98.64% of the 30025 tools)\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# nb of tools with topic\n",
    "\n",
    "SELECT (COUNT(DISTINCT ?tool) AS ?nbTools)\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:applicationSubCategory ?topic .\n",
    "  ?topic rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?topic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+query)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "\n",
    "nbToolsWithTopic = int(results[\"results\"][\"bindings\"][0][\"nbTools\"][\"value\"])\n",
    "print(\"Nb tools with topic: {} ({:.2f}% of the {} tools)\".format(nbToolsWithTopic, nbToolsWithTopic*100/nbTools, nbTools))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18e89147",
   "metadata": {},
   "source": [
    "### Number of tools with at least one operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "755ae4a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb tools with operation: 28299 (94.25% of the 30025 tools)\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# nb of tools with operation\n",
    "\n",
    "SELECT (COUNT(DISTINCT ?tool) AS ?nbTools)\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:featureList ?operation .\n",
    "  ?operation rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?operation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+query)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "\n",
    "nbToolsWithOperation = int(results[\"results\"][\"bindings\"][0][\"nbTools\"][\"value\"])\n",
    "print(\"Nb tools with operation: {} ({:.2f}% of the {} tools)\".format(nbToolsWithOperation, nbToolsWithOperation*100/nbTools, nbTools))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac60d66c",
   "metadata": {},
   "source": [
    "### List of topics directly assigned to tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f4374e03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>topic</th>\n",
       "      <th>topicLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0780</td>\n",
       "      <td>Plant biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_2885</td>\n",
       "      <td>DNA polymorphism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0659</td>\n",
       "      <td>Functional, regulatory and non-coding RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0203</td>\n",
       "      <td>Gene expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0199</td>\n",
       "      <td>Genetic variation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       tool                               topic  \\\n",
       "0  https://bio.tools/pmirkb  http://edamontology.org/topic_0780   \n",
       "1  https://bio.tools/pmirkb  http://edamontology.org/topic_2885   \n",
       "2  https://bio.tools/pmirkb  http://edamontology.org/topic_0659   \n",
       "3  https://bio.tools/pmirkb  http://edamontology.org/topic_0203   \n",
       "4  https://bio.tools/pmirkb  http://edamontology.org/topic_0199   \n",
       "\n",
       "                                  topicLabel  \n",
       "0                              Plant biology  \n",
       "1                           DNA polymorphism  \n",
       "2  Functional, regulatory and non-coding RNA  \n",
       "3                            Gene expression  \n",
       "4                          Genetic variation  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# tools topics\n",
    "\n",
    "SELECT DISTINCT ?tool ?topic ?topicLabel\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:applicationSubCategory ?topic .\n",
    "  ?topic rdf:type owl:Class . # for avoiding deprecated topics\n",
    "  FILTER NOT EXISTS { ?topic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?topic rdfs:label ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?topicLabel)\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "dfToolTopic = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "dfToolTopic.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3f8c26",
   "metadata": {},
   "source": [
    "### List of topics directly assigned and inherited to tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d82aaeef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>topic</th>\n",
       "      <th>topicLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0780</td>\n",
       "      <td>Plant biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_3070</td>\n",
       "      <td>Biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0003</td>\n",
       "      <td>Topic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_2885</td>\n",
       "      <td>DNA polymorphism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/topic_0199</td>\n",
       "      <td>Genetic variation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       tool                               topic  \\\n",
       "0  https://bio.tools/pmirkb  http://edamontology.org/topic_0780   \n",
       "1  https://bio.tools/pmirkb  http://edamontology.org/topic_3070   \n",
       "2  https://bio.tools/pmirkb  http://edamontology.org/topic_0003   \n",
       "3  https://bio.tools/pmirkb  http://edamontology.org/topic_2885   \n",
       "4  https://bio.tools/pmirkb  http://edamontology.org/topic_0199   \n",
       "\n",
       "          topicLabel  \n",
       "0      Plant biology  \n",
       "1            Biology  \n",
       "2              Topic  \n",
       "3   DNA polymorphism  \n",
       "4  Genetic variation  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# tools topics (and ancestors)\n",
    "\n",
    "SELECT DISTINCT ?tool ?topic ?topicLabel\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:applicationSubCategory/(rdfs:subClassOf*) ?topic .\n",
    "  ?topic rdf:type owl:Class . # for avoiding deprecated topics and owl:Restriction\n",
    "  FILTER NOT EXISTS { ?topic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?topic rdfs:label ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?topicLabel)\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "dfToolTopicTransitive = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "dfToolTopicTransitive.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92cf53d6",
   "metadata": {},
   "source": [
    "### List of operations directly assigned to tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6ec10578",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>operation</th>\n",
       "      <th>operationLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2428</td>\n",
       "      <td>Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0484</td>\n",
       "      <td>SNP detection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0308</td>\n",
       "      <td>PCR primer design</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2409</td>\n",
       "      <td>Data handling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0463</td>\n",
       "      <td>miRNA target prediction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       tool                               operation  \\\n",
       "0  https://bio.tools/pmirkb  http://edamontology.org/operation_2428   \n",
       "1  https://bio.tools/pmirkb  http://edamontology.org/operation_0484   \n",
       "2  https://bio.tools/pmirkb  http://edamontology.org/operation_0308   \n",
       "3  https://bio.tools/pmirkb  http://edamontology.org/operation_2409   \n",
       "4  https://bio.tools/pmirkb  http://edamontology.org/operation_0463   \n",
       "\n",
       "            operationLabel  \n",
       "0               Validation  \n",
       "1            SNP detection  \n",
       "2        PCR primer design  \n",
       "3            Data handling  \n",
       "4  miRNA target prediction  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# tools operations\n",
    "\n",
    "SELECT DISTINCT ?tool ?operation ?operationLabel\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:featureList ?operation .\n",
    "  ?operation rdf:type owl:Class . # for avoiding deprecated operations\n",
    "  FILTER NOT EXISTS { ?operation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?operation rdfs:label ?oLabel }\n",
    "  BIND(COALESCE(?oLabel, \"\") AS ?operationLabel)\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "dfToolOperation = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "dfToolOperation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b62373a5",
   "metadata": {},
   "source": [
    "### List of topics directly assigned and inherited to tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a120de46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>operation</th>\n",
       "      <th>operationLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2428</td>\n",
       "      <td>Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0004</td>\n",
       "      <td>Operation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_0484</td>\n",
       "      <td>SNP detection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_3227</td>\n",
       "      <td>Variant calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>http://edamontology.org/operation_2478</td>\n",
       "      <td>Nucleic acid sequence analysis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       tool                               operation  \\\n",
       "0  https://bio.tools/pmirkb  http://edamontology.org/operation_2428   \n",
       "1  https://bio.tools/pmirkb  http://edamontology.org/operation_0004   \n",
       "2  https://bio.tools/pmirkb  http://edamontology.org/operation_0484   \n",
       "3  https://bio.tools/pmirkb  http://edamontology.org/operation_3227   \n",
       "4  https://bio.tools/pmirkb  http://edamontology.org/operation_2478   \n",
       "\n",
       "                   operationLabel  \n",
       "0                      Validation  \n",
       "1                       Operation  \n",
       "2                   SNP detection  \n",
       "3                 Variant calling  \n",
       "4  Nucleic acid sequence analysis  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# tools operations (and ancestors)\n",
    "\n",
    "SELECT DISTINCT ?tool ?operation ?operationLabel\n",
    "WHERE {\n",
    "  ?tool rdf:type sc:SoftwareApplication .\n",
    "  ?tool sc:featureList/(rdfs:subClassOf*) ?operation . # for avoiding deprecated operations and owl:Restriction\n",
    "  ?operation rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?operation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  OPTIONAL { ?operation rdfs:label ?oLabel }\n",
    "  BIND(COALESCE(?oLabel, \"\") AS ?operationLabel)\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "dfToolOperationTransitive = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "dfToolOperationTransitive.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb573bef",
   "metadata": {},
   "source": [
    "## Create Dataframe for Figure 2 : dfTool with directly assigned annotations and inherited annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00f9fec0",
   "metadata": {},
   "source": [
    "#### Create dataframe with directly assigned annotations Topics and Operations and theirs numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d395806a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>toolLabel</th>\n",
       "      <th>nbTopics</th>\n",
       "      <th>nbOperations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>PmiRKB</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/PCA-PAM50</td>\n",
       "      <td>PCA-PAM50</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/CEthreader</td>\n",
       "      <td>CEthreader</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/ipath</td>\n",
       "      <td>iPATH</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/nidm</td>\n",
       "      <td>NIDM</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tool   toolLabel  nbTopics  nbOperations\n",
       "0      https://bio.tools/pmirkb      PmiRKB         5             5\n",
       "1   https://bio.tools/PCA-PAM50   PCA-PAM50         3             2\n",
       "2  https://bio.tools/CEthreader  CEthreader         5             5\n",
       "3       https://bio.tools/ipath       iPATH         2             0\n",
       "4        https://bio.tools/nidm        NIDM         5             4"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfTool = pd.read_csv(\"../Dataframes/dfTool.tsv.bz2\", sep=\"\\t\")\n",
    "dfToolNbTopics = dfToolTopic.groupby(by='tool').size().reset_index(name='nbTopics').sort_values(by=\"nbTopics\", ascending=False)\n",
    "dfToolNbOperations = dfToolOperation.groupby(by='tool').size().reset_index(name='nbOperations').sort_values(by=\"nbOperations\", ascending=False)\n",
    "\n",
    "dfTool = dfTool.join(dfToolNbTopics.set_index('tool'), on='tool')\n",
    "dfTool = dfTool.join(dfToolNbOperations.set_index('tool'), on='tool')\n",
    "dfTool['nbTopics'] = dfTool['nbTopics'].fillna(0)\n",
    "dfTool['nbOperations'] = dfTool['nbOperations'].fillna(0)\n",
    "dfTool = dfTool.astype({'nbTopics':'int'})\n",
    "dfTool = dfTool.astype({'nbOperations':'int'})\n",
    "dfTool_NT = dfTool\n",
    "dfTool_NT.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17a59afb",
   "metadata": {},
   "source": [
    "#### Create dataframe with inherited annotations Topics and Operations and theirs numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8a59feea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>toolLabel</th>\n",
       "      <th>nbTopics</th>\n",
       "      <th>nbOperations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>PmiRKB</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/PCA-PAM50</td>\n",
       "      <td>PCA-PAM50</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/CEthreader</td>\n",
       "      <td>CEthreader</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/ipath</td>\n",
       "      <td>iPATH</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/nidm</td>\n",
       "      <td>NIDM</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tool   toolLabel  nbTopics  nbOperations\n",
       "0      https://bio.tools/pmirkb      PmiRKB        14            18\n",
       "1   https://bio.tools/PCA-PAM50   PCA-PAM50         7            11\n",
       "2  https://bio.tools/CEthreader  CEthreader        12            23\n",
       "3       https://bio.tools/ipath       iPATH         6             0\n",
       "4        https://bio.tools/nidm        NIDM        12            12"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfTool = pd.read_csv(\"../Dataframes/dfTool.tsv.bz2\", sep=\"\\t\")\n",
    "dfToolNbTopics = dfToolTopicTransitive.groupby(by='tool').size().reset_index(name='nbTopics').sort_values(by=\"nbTopics\", ascending=False)\n",
    "dfToolNbOperations = dfToolOperationTransitive.groupby(by='tool').size().reset_index(name='nbOperations').sort_values(by=\"nbOperations\", ascending=False)\n",
    "\n",
    "dfTool = dfTool.join(dfToolNbTopics.set_index('tool'), on='tool')\n",
    "dfTool = dfTool.join(dfToolNbOperations.set_index('tool'), on='tool')\n",
    "dfTool['nbTopics'] = dfTool['nbTopics'].fillna(0)\n",
    "dfTool['nbOperations'] = dfTool['nbOperations'].fillna(0)\n",
    "dfTool = dfTool.astype({'nbTopics':'int'})\n",
    "dfTool = dfTool.astype({'nbOperations':'int'})\n",
    "dfTool_T = dfTool\n",
    "dfTool_T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85b4448f",
   "metadata": {},
   "source": [
    "## Create Dataframe for Figure 3 : dfToolallmetrics for inherited annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a86e1f9",
   "metadata": {},
   "source": [
    "### Inherited Topics, metrics and compute of IC and entropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b799cea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb topics (incl. ancestors): 258\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>topic</th>\n",
       "      <th>topicLabel</th>\n",
       "      <th>nbTools</th>\n",
       "      <th>frequence</th>\n",
       "      <th>IC</th>\n",
       "      <th>entropy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://edamontology.org/topic_0780</td>\n",
       "      <td>Plant biology</td>\n",
       "      <td>882</td>\n",
       "      <td>0.029781</td>\n",
       "      <td>5.069454</td>\n",
       "      <td>0.150974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://edamontology.org/topic_3070</td>\n",
       "      <td>Biology</td>\n",
       "      <td>18273</td>\n",
       "      <td>0.616998</td>\n",
       "      <td>0.696663</td>\n",
       "      <td>0.429840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>http://edamontology.org/topic_0003</td>\n",
       "      <td>Topic</td>\n",
       "      <td>29616</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>http://edamontology.org/topic_2885</td>\n",
       "      <td>DNA polymorphism</td>\n",
       "      <td>1202</td>\n",
       "      <td>0.040586</td>\n",
       "      <td>4.622868</td>\n",
       "      <td>0.187625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>http://edamontology.org/topic_0199</td>\n",
       "      <td>Genetic variation</td>\n",
       "      <td>2907</td>\n",
       "      <td>0.098156</td>\n",
       "      <td>3.348774</td>\n",
       "      <td>0.328704</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                topic         topicLabel  nbTools  frequence  \\\n",
       "0  http://edamontology.org/topic_0780      Plant biology      882   0.029781   \n",
       "1  http://edamontology.org/topic_3070            Biology    18273   0.616998   \n",
       "2  http://edamontology.org/topic_0003              Topic    29616   1.000000   \n",
       "3  http://edamontology.org/topic_2885   DNA polymorphism     1202   0.040586   \n",
       "4  http://edamontology.org/topic_0199  Genetic variation     2907   0.098156   \n",
       "\n",
       "         IC   entropy  \n",
       "0  5.069454  0.150974  \n",
       "1  0.696663  0.429840  \n",
       "2 -0.000000 -0.000000  \n",
       "3  4.622868  0.187625  \n",
       "4  3.348774  0.328704  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbTopicsTransitive = dfToolTopicTransitive['topic'].nunique()\n",
    "print(\"Nb topics (incl. ancestors): {}\".format(nbTopicsTransitive))\n",
    "dfTopicNbTools = dfToolTopicTransitive[['tool', 'topic']].groupby(by='topic').size().reset_index(name='nbTools').sort_values(by=\"nbTools\", ascending=False)\n",
    "dfTopic = dfToolTopicTransitive[['topic', 'topicLabel']].drop_duplicates(subset=['topic', 'topicLabel'], keep='first').reset_index(drop=True)\n",
    "dfTopic = dfTopic.join(dfTopicNbTools.set_index('topic'), on='topic')\n",
    "dfTopic['frequence'] = dfTopic['nbTools'] / nbToolsWithTopic\n",
    "dfTopic['IC'] = -np.log2(dfTopic['frequence'])\n",
    "dfTopic['entropy'] = dfTopic['frequence']*dfTopic['IC']\n",
    "dfTopic.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d60516c0",
   "metadata": {},
   "source": [
    "### Inherited Operations, metrics and compute of IC and entropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3d92d7c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb operations (incl. ancestors): 527\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>operation</th>\n",
       "      <th>operationLabel</th>\n",
       "      <th>nbTools</th>\n",
       "      <th>frequence</th>\n",
       "      <th>IC</th>\n",
       "      <th>entropy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://edamontology.org/operation_2428</td>\n",
       "      <td>Validation</td>\n",
       "      <td>1307</td>\n",
       "      <td>0.046185</td>\n",
       "      <td>4.436420</td>\n",
       "      <td>0.204898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://edamontology.org/operation_0004</td>\n",
       "      <td>Operation</td>\n",
       "      <td>28299</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>http://edamontology.org/operation_0484</td>\n",
       "      <td>SNP detection</td>\n",
       "      <td>330</td>\n",
       "      <td>0.011661</td>\n",
       "      <td>6.422141</td>\n",
       "      <td>0.074890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>http://edamontology.org/operation_3227</td>\n",
       "      <td>Variant calling</td>\n",
       "      <td>1267</td>\n",
       "      <td>0.044772</td>\n",
       "      <td>4.481263</td>\n",
       "      <td>0.200635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>http://edamontology.org/operation_2478</td>\n",
       "      <td>Nucleic acid sequence analysis</td>\n",
       "      <td>7821</td>\n",
       "      <td>0.276370</td>\n",
       "      <td>1.855326</td>\n",
       "      <td>0.512757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                operation                  operationLabel  \\\n",
       "0  http://edamontology.org/operation_2428                      Validation   \n",
       "1  http://edamontology.org/operation_0004                       Operation   \n",
       "2  http://edamontology.org/operation_0484                   SNP detection   \n",
       "3  http://edamontology.org/operation_3227                 Variant calling   \n",
       "4  http://edamontology.org/operation_2478  Nucleic acid sequence analysis   \n",
       "\n",
       "   nbTools  frequence        IC   entropy  \n",
       "0     1307   0.046185  4.436420  0.204898  \n",
       "1    28299   1.000000 -0.000000 -0.000000  \n",
       "2      330   0.011661  6.422141  0.074890  \n",
       "3     1267   0.044772  4.481263  0.200635  \n",
       "4     7821   0.276370  1.855326  0.512757  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbOperationsTransitive = dfToolOperationTransitive['operation'].nunique()\n",
    "print(\"Nb operations (incl. ancestors): {}\".format(nbOperationsTransitive))\n",
    "dfOperationNbTools = dfToolOperationTransitive[['tool', 'operation']].groupby(by='operation').size().reset_index(name='nbTools').sort_values(by=\"nbTools\", ascending=False)\n",
    "dfOperation = dfToolOperationTransitive[['operation', 'operationLabel']].drop_duplicates(subset=['operation', 'operationLabel'], keep='first').reset_index(drop=True)\n",
    "dfOperation = dfOperation.join(dfOperationNbTools.set_index('operation'), on='operation')\n",
    "dfOperation['frequence'] = dfOperation['nbTools'] / nbToolsWithOperation\n",
    "dfOperation['IC'] = -np.log2(dfOperation['frequence'])\n",
    "dfOperation['entropy'] = dfOperation['frequence']*dfOperation['IC']\n",
    "dfOperation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be298391",
   "metadata": {},
   "source": [
    "### Compute IC and entropy for tools with inherited annotations, creation of dfToolallmetrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a13e522d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>toolLabel</th>\n",
       "      <th>nbTopics</th>\n",
       "      <th>nbOperations</th>\n",
       "      <th>topicScore</th>\n",
       "      <th>operationScore</th>\n",
       "      <th>score</th>\n",
       "      <th>topicEntropy</th>\n",
       "      <th>operationEntropy</th>\n",
       "      <th>entropy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://bio.tools/pmirkb</td>\n",
       "      <td>PmiRKB</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>35.012864</td>\n",
       "      <td>70.687577</td>\n",
       "      <td>105.700441</td>\n",
       "      <td>4.743019</td>\n",
       "      <td>4.103735</td>\n",
       "      <td>8.846754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://bio.tools/PCA-PAM50</td>\n",
       "      <td>PCA-PAM50</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>19.418860</td>\n",
       "      <td>32.575948</td>\n",
       "      <td>51.994807</td>\n",
       "      <td>1.906146</td>\n",
       "      <td>3.104192</td>\n",
       "      <td>5.010339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://bio.tools/CEthreader</td>\n",
       "      <td>CEthreader</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>37.561111</td>\n",
       "      <td>92.221264</td>\n",
       "      <td>129.782375</td>\n",
       "      <td>3.412990</td>\n",
       "      <td>5.632084</td>\n",
       "      <td>9.045074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://bio.tools/ipath</td>\n",
       "      <td>iPATH</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>14.724993</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.724993</td>\n",
       "      <td>1.719522</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.719522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://bio.tools/nidm</td>\n",
       "      <td>NIDM</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>41.797858</td>\n",
       "      <td>42.837473</td>\n",
       "      <td>84.635331</td>\n",
       "      <td>3.539804</td>\n",
       "      <td>2.743541</td>\n",
       "      <td>6.283344</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tool   toolLabel  nbTopics  nbOperations  \\\n",
       "0      https://bio.tools/pmirkb      PmiRKB        14            18   \n",
       "1   https://bio.tools/PCA-PAM50   PCA-PAM50         7            11   \n",
       "2  https://bio.tools/CEthreader  CEthreader        12            23   \n",
       "3       https://bio.tools/ipath       iPATH         6             0   \n",
       "4        https://bio.tools/nidm        NIDM        12            12   \n",
       "\n",
       "   topicScore  operationScore       score  topicEntropy  operationEntropy  \\\n",
       "0   35.012864       70.687577  105.700441      4.743019          4.103735   \n",
       "1   19.418860       32.575948   51.994807      1.906146          3.104192   \n",
       "2   37.561111       92.221264  129.782375      3.412990          5.632084   \n",
       "3   14.724993        0.000000   14.724993      1.719522          0.000000   \n",
       "4   41.797858       42.837473   84.635331      3.539804          2.743541   \n",
       "\n",
       "    entropy  \n",
       "0  8.846754  \n",
       "1  5.010339  \n",
       "2  9.045074  \n",
       "3  1.719522  \n",
       "4  6.283344  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dfToolTopicTransitive.join(dfTopic[['topic', 'IC', 'entropy']].set_index('topic'), on='topic')[['tool', 'IC']].groupby(by='tool').sum().rename(columns={'IC': 'topicScore'}).reset_index('tool')\n",
    "dfTool = dfTool.join(df.set_index('tool'), on='tool')\n",
    "dfTool['topicScore'] = dfTool['topicScore'].fillna(0)\n",
    "\n",
    "df = dfToolOperationTransitive.join(dfOperation[['operation', 'IC', 'entropy']].set_index('operation'), on='operation')[['tool', 'IC']].groupby(by='tool').sum().rename(columns={'IC': 'operationScore'}).reset_index('tool')\n",
    "dfTool = dfTool.join(df.set_index('tool'), on='tool')\n",
    "dfTool['operationScore'] = dfTool['operationScore'].fillna(0)\n",
    "\n",
    "dfTool['score'] = dfTool['topicScore'] + dfTool['operationScore']\n",
    "\n",
    "\n",
    "df = dfToolTopicTransitive.join(dfTopic[['topic', 'IC', 'entropy']].set_index('topic'), on='topic')[['tool', 'entropy']].groupby(by='tool').sum().rename(columns={'entropy': 'topicEntropy'}).reset_index('tool')\n",
    "dfTool = dfTool.join(df.set_index('tool'), on='tool')\n",
    "dfTool['topicEntropy'] = dfTool['topicEntropy'].fillna(0)\n",
    "\n",
    "df = dfToolOperationTransitive.join(dfOperation[['operation', 'IC', 'entropy']].set_index('operation'), on='operation')[['tool', 'entropy']].groupby(by='tool').sum().rename(columns={'entropy': 'operationEntropy'}).reset_index('tool')\n",
    "dfTool = dfTool.join(df.set_index('tool'), on='tool')\n",
    "dfTool['operationEntropy'] = dfTool['operationEntropy'].fillna(0)\n",
    "\n",
    "dfTool['entropy'] = dfTool['topicEntropy'] + dfTool['operationEntropy']\n",
    "dfToolallmetrics = dfTool\n",
    "dfToolallmetrics.head()\n",
    "\n",
    "#Can generate dfToolallmetrics.tsv.bz2 : \n",
    "#dfToolallmetrics.to_csv(\"../Dataframes/dfToolallmetrics.tsv.bz2\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fce4125",
   "metadata": {},
   "source": [
    "### Explanation of calculated scores \n",
    "- topicScore: Score based on the IC (Information Content) of topics associated with a tool.\n",
    "- operationScore: Score based on the IC of operations associated with a tool.\n",
    "- score: Sum of the two previous scores.\n",
    "- topicEntropy: Total entropy of topics associated with a tool.\n",
    "- operationEntropy: Total entropy of operations associated with a tool.\n",
    "- entropy: Sum of the previous entropies."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04d4c590",
   "metadata": {},
   "source": [
    "## Redundancy of tools "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a7a524b",
   "metadata": {},
   "source": [
    "### Topic redundancy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4418a39b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3405'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redundancyQuery = \"\"\"\n",
    "SELECT (COUNT(DISTINCT ?tool) AS ?nbToolsWithRedundantTopics)\n",
    "WHERE {\n",
    "  ?tool sc:applicationSubCategory ?redundantDirectTopic .\n",
    "  ?redundantDirectTopic rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?redundantDirectTopic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  \n",
    "  ?tool sc:applicationSubCategory ?directTopic .\n",
    "  ?directTopic rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?directTopic rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  \n",
    "  ?directTopic rdfs:subClassOf+ ?redundantDirectTopic .\n",
    "  \n",
    "  OPTIONAL { ?directTopic rdfs:label ?tLabel }\n",
    "  BIND(COALESCE(?tLabel, \"\") AS ?directTopicLabel)\n",
    "  \n",
    "  OPTIONAL { ?redundantDirectTopic rdfs:label ?rtLabel }\n",
    "  BIND(COALESCE(?rtLabel, \"\") AS ?redundantDirectTopicLabel)\n",
    "}\n",
    "#LIMIT 10\n",
    "\"\"\"\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+redundancyQuery)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "results[\"results\"][\"bindings\"][0][\"nbToolsWithRedundantTopics\"][\"value\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aa70754",
   "metadata": {},
   "source": [
    "### Operation redundancy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "254a11b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2055'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redundancyQuery = \"\"\"\n",
    "SELECT (COUNT(DISTINCT ?tool) AS ?nbToolsWithRedundantOperations)\n",
    "WHERE {\n",
    "  ?tool sc:featureList ?redundantDirectOperation .\n",
    "  ?redundantDirectOperation rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?redundantDirectOperation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  \n",
    "  ?tool sc:featureList ?directOperation .\n",
    "  ?directOperation rdf:type owl:Class .\n",
    "  FILTER NOT EXISTS { ?directOperation rdfs:subClassOf? owl:DeprecatedClass }\n",
    "  \n",
    "  ?directOperation rdfs:subClassOf+ ?redundantDirectOperation .\n",
    "  \n",
    "  OPTIONAL { ?directOperation rdfs:label ?dLabel }\n",
    "  BIND(COALESCE(?dLabel, \"\") AS ?directOperationLabel)\n",
    "  \n",
    "  OPTIONAL { ?redundantDirectOperation rdfs:label ?rdLabel }\n",
    "  BIND(COALESCE(?rdLabel, \"\") AS ?redundantDirectOperationLabel)\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+redundancyQuery)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "results[\"results\"][\"bindings\"][0][\"nbToolsWithRedundantOperations\"][\"value\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79341a83",
   "metadata": {},
   "source": [
    "3405 tools have a redundant topic annotation on 30 025\n",
    "\n",
    "2055 tools have a redundant operation annotation on 30 325 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dada3cf",
   "metadata": {},
   "source": [
    "## Deprecated annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7f0935ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><tr><th>nbDeprecatedItem</th></tr><tr><td>1114</td></tr></table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "# deprecated items\n",
    "\n",
    "SELECT (COUNT(DISTINCT ?deprecatedItem) AS ?nbDeprecatedItem) #?topic\n",
    "#SELECT DISTINCT ?deprecatedItem\n",
    "WHERE {\n",
    "  { ?deprecatedItem rdfs:subClassOf owl:DeprecatedClass }\n",
    "  UNION\n",
    "  { ?deprecatedItem owl:deprecated true . }\n",
    "  UNION\n",
    "  { ?deprecatedItem owl:deprecated \"true\" . }\n",
    "  UNION \n",
    "  { ?deprecatedItem owl:deprecated \"True\" . }\n",
    "  \n",
    "}\n",
    "ORDER BY ?deprecatedItem\n",
    "\"\"\"\n",
    "\n",
    "#df = sparqldataframe.query(endpointURL, prefixes+query)\n",
    "#df\n",
    "\n",
    "sparql = SPARQLWrapper(endpointURL)\n",
    "sparql.setQuery(prefixes+query)\n",
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n",
    "displaySparqlResults(results)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biotoolsEDAMvisualization",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
